<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="APL is a package developed for computation of Association Plots
  (AP), a method for visualization and analysis of single cell transcriptomics 
  data. The main focus of APL is the identification of genes characteristic for
  individual clusters of cells from input data. The package performs 
  correspondence analysis (CA) and allows to identify cluster-specific
  genes using Association Plots. Additionally, APL computes the 
  cluster-specificity scores for all genes which allows to rank the genes by 
  their specificity for a selected cell cluster of interest.">
<title>Association Plots â€¢ APL</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="Association Plots">
<meta property="og:description" content="APL is a package developed for computation of Association Plots
  (AP), a method for visualization and analysis of single cell transcriptomics 
  data. The main focus of APL is the identification of genes characteristic for
  individual clusters of cells from input data. The package performs 
  correspondence analysis (CA) and allows to identify cluster-specific
  genes using Association Plots. Additionally, APL computes the 
  cluster-specificity scores for all genes which allows to rank the genes by 
  their specificity for a selected cell cluster of interest.">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="index.html">APL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="articles/APL.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9">
<p><img src="reference/figures/fig_AP.png" width="700"></p>
<div class="section level1">
<div class="page-header"><h1 id="apl">APL<a class="anchor" aria-label="anchor" href="#apl"></a>
</h1></div>
<p><code>APL</code> is a package developed for computation of Association Plots, a method for visualization and analysis of single cell transcriptomics data. The main focus of <code>APL</code> is the identification of genes characteristic for individual clusters of cells from input data.</p>
<p>When working with <code>APL</code> package please cite:</p>
<pre><code>Gralinska, E., Kohl, C., Fadakar, B. S., &amp; Vingron, M. (2022). 
Visualizing Cluster-specific Genes from Single-cell Transcriptomics Data Using Association Plots. 
Journal of Molecular Biology, 434(11), 167525.</code></pre>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>The <code>APL</code> can be installed from GitHub:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(devtools)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install_github</span>(<span class="st">"VingronLab/APL"</span>)</span></code></pre></div>
<p>To additionally build the package vignette, run instead:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install_github</span>(<span class="st">"VingronLab/APL"</span>, <span class="at">build_vignettes =</span> <span class="cn">TRUE</span>, <span class="at">dependencies =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>Building the vignette will however take considerable time.</p>
<p><strong>The vignette can also be found under the link: <a href="https://vingronlab.github.io/APL/" class="uri">https://vingronlab.github.io/APL/</a> (hyperlink in the GitHub repository description).</strong></p>
<p>To install the <code>APL</code> from Bioconductor, run:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>))</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">"BiocManager"</span>)</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">"APL"</span>)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="pytorch-installation">Pytorch installation<a class="anchor" aria-label="anchor" href="#pytorch-installation"></a>
</h2>
<p>In order to speed up the singular value decomposition, we highly recommend the installation of <code>pytorch</code>. Users can instead also opt to use the slower R native SVD. For this, please set the argument <code>python = FALSE</code> wherever applicable in the package vignette.</p>
<div class="section level3">
<h3 id="install-pytorch-with-reticulate">Install pytorch with reticulate<a class="anchor" aria-label="anchor" href="#install-pytorch-with-reticulate"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(reticulate)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">install_miniconda</span>() </span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">conda_install</span>(<span class="at">envname =</span> <span class="st">"r-reticulate"</span>, <span class="at">packages =</span> <span class="st">"numpy"</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="fu">conda_install</span>(<span class="at">envname =</span> <span class="st">"r-reticulate"</span>, <span class="at">packages =</span> <span class="st">"pytorch"</span>)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="manually-install-pytorch-with-conda">Manually install pytorch with conda<a class="anchor" aria-label="anchor" href="#manually-install-pytorch-with-conda"></a>
</h3>
<p>Download the appropriate Miniconda installer for your system from <a href="https://docs.conda.io/en/latest/miniconda.html" class="external-link">the conda website</a>. Follow the installation instructions on their website and make sure the R package <code>reticulate</code> is also installed before proceeding. Once installed, list all available conda environments via <br><code>conda info --envs</code> <br> One of the environments should have <code>r-reticulate</code> in its name. Depending on where you installed it and your system, the exact path might be different. Activate the environment and install pytorch into it.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>conda activate <span class="sc">~</span><span class="er">/</span>.local<span class="sc">/</span>share<span class="sc">/</span>r<span class="sc">-</span>miniconda<span class="sc">/</span>envs<span class="sc">/</span>r<span class="sc">-</span>reticulate <span class="co"># change path accordingly.</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>conda install numpy</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>conda install pytorch</span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="feature-overview">Feature overview<a class="anchor" aria-label="anchor" href="#feature-overview"></a>
</h2>
<p>Please run</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="fu">vignette</span>(<span class="st">"APL"</span>)</span></code></pre></div>
<p>after installation with <code>build_vignettes = TRUE</code> for an introduction into the package.</p>
</div>
</div>

  </main><aside class="col-md-3"><div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://www.bioconductor.org/packages/APL" class="external-link">View on Bioconductor</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing APL</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Clemens Kohl <br><small class="roles"> Maintainer, author </small>  </li>
<li>Elzbieta Gralinska <br><small class="roles"> Author </small>  </li>
<li>Martin Vingron <br><small class="roles"> Author </small>  </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://bioconductor.org/checkResults/release/bioc-LATEST/APL" class="external-link"><img src="http://www.bioconductor.org/shields/build/release/bioc/APL.svg" alt="Bioc release status"></a></li>
<li><a href="https://bioconductor.org/checkResults/devel/bioc-LATEST/APL" class="external-link"><img src="http://www.bioconductor.org/shields/build/devel/bioc/APL.svg" alt="Bioc devel status"></a></li>
<li><a href="http://bioconductor.org/packages/stats/bioc/APL/" class="external-link"><img src="https://bioconductor.org/shields/downloads/release/APL.svg" alt="Bioc downloads rank"></a></li>
<li><a href="https://support.bioconductor.org/tag/APL" class="external-link"><img src="https://bioconductor.org/shields/posts/APL.svg" alt="Bioc support"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/APL.html#since" class="external-link"><img src="https://bioconductor.org/shields/years-in-bioc/APL.svg" alt="Bioc history"></a></li>
<li><a href="http://bioconductor.org/checkResults/devel/bioc-LATEST/APL/" class="external-link"><img src="https://bioconductor.org/shields/lastcommit/devel/bioc/APL.svg" alt="Bioc last commit"></a></li>
<li><a href="https://bioconductor.org/packages/release/bioc/html/APL.html#since" class="external-link"><img src="https://bioconductor.org/shields/dependencies/release/APL.svg" alt="Bioc dependencies"></a></li>
</ul>
</div>

  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Clemens Kohl, Elzbieta Gralinska, Martin Vingron.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
