---
title: "Analyzing data with APL"
author:
- name: Elzbieta Gralinska
  affiliation: Max Planck Institute for Molecular Genetics
  email: gralinska@molgen.mpg.de
- name: Clemens Kohl
  affiliation: Max Planck Institute for Molecular Genetics
  email: kohl@molgen.mpg.de
- name: Martin Vingron Kohl
  affiliation: Max Planck Institute for Molecular Genetics
  email: vingron@molgen.mpg.de
package: APL
output:
  BiocStyle::html_document
abstract: |
  Description of your vignette
vignette: |
  %\VignetteIndexEntry{Analyzing data with APL}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```
  
# Introduction

About APL, Association Plots, finding cluster-specific genes, etc.


When working with `r Rpackage("APL")` package please cite:
```
Association Plots: Visualizing associations in high-dimensional correspondence analysis biplots
Elzbieta Gralinska, Martin Vingron
bioRxiv 2020.10.23.352096; doi: https://doi.org/10.1101/2020.10.23.352096
```

For mathematical description of the method, please refer to the manuscript.

# Pytorch installation
In order to speed up the singular value decomposition, we highly recommend the installation of pytorch.
Users can instead also opt to instead use the slower R native SVD. For this please turn the argument `python = FALSE` wherever applicable in this vignette.


## Install pytorch with reticulate
```{r reticulate, eval=FALSE}
library(reticulate)
install_miniconda() 
conda_install(envname = "r-reticulate", packages = "numpy")
conda_install(envname = "r-reticulate", packages = "pytorch")
```

## Manually install pytorch with conda
Download the appropriate Miniconda installer for your system from [the conda website](https://docs.conda.io/en/latest/miniconda.html). 
Follow the installation instructions on their website and make sure the R package reticulate is also installed before proceeding.
Once installed, list all available conda environments via <br>
`conda info --envs` <br>
One of the environments should have `r-reticulate` in its name. Depending on where
you installed it and your system, the exact path might be different.
Activate the environment and install pytorch into it.

```{bash conda, eval=FALSE}
conda activate ~/.local/share/r-miniconda/envs/r-reticulate # change path accordingly.
conda install numpy
conda install pytorch
```


# Preprocessing
## Setup
For this vignette we will use peripheral blood mononuclear cell (PBMC) data from 10x Genomics.
The data necessary to follow the vignette can be downloaded from [here](https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz).
Besides the package `r Rpackage("APL")` we will be using the single-cell RNAseq analysis suite `r CRANpkg("Seurat")` (V. 4.0.4) to preprocess the data, but the preprocessing could equally be done with `r Biocpkg("SingleCellExperiment")` and `r Biocpkg("scater")`\\`r Biocpkg("scran")`.
For the preprocessing we follow the [Guided Clustering Tutorial](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html) from the Seurat vignettes. For details about the data preprocessing please to refer to their website.

```{r setup, message=FALSE, warning=FALSE}
library(APL)
library(Seurat)
library(ggplot2)
set.seed(1234)

```

## Loading the data
**CHANGE FILE PATH OR FIND OTHER SOLUTION**
```{r load_data, message=FALSE, warning=FALSE}
# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "/project/kohl_data/datasets/pbmc3k/filtered_gene_bc_matrices/hg19/")

# Remove genes not expressed in any gene.
no_zeros_rows <- rowSums(pbmc.data) > 0
pbmc.data <- pbmc.data[no_zeros_rows,]

# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)

pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

# filter data
pbmc <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)

```

## Normalization, PCA & Clustering
```{r preprocessing}

# normalization
pbmc <- NormalizeData(pbmc,
                      normalization.method = "LogNormalize",
                      scale.factor = 10000,
                      verbose = FALSE)

pbmc <- FindVariableFeatures(pbmc,
                             selection.method = "vst",
                             nfeatures = 2000,
                             verbose = FALSE)

# scaling
all.genes <- rownames(pbmc)
pbmc <- ScaleData(pbmc,
                  features = all.genes,
                  verbose = FALSE)

# Run PCA
pbmc <- RunPCA(pbmc,
               features = VariableFeatures(object = pbmc),
               verbose = FALSE)

# Cluster cells
pbmc <- FindNeighbors(pbmc, dims = 1:10, verbose = FALSE)
pbmc <- FindClusters(pbmc, resolution = 0.5, verbose = FALSE)

pbmc <- RunUMAP(pbmc, dims = 1:10, verbose = FALSE)

new.cluster.ids <- c("0 - Naive CD4 T",
                     "1 - CD14+ Mono", 
                     "2 - Memory CD4 T", 
                     "3 - B", "4 - CD8 T", 
                     "5 - FCGR3A+ Mono", 
                     "6 - NK", "7 - DC", 
                     "8 - Platelet")

names(new.cluster.ids) <- levels(pbmc)
pbmc <- RenameIdents(pbmc, new.cluster.ids)
pbmc$cell_type <- Idents(pbmc)

DimPlot(pbmc, reduction = "umap", label = F, pt.size = 0.5)
```

# Correspondence Analysis
In a first step we perform Correspondence Analysis (CA) on the log normalized 
counts from the data. CA is a method for dimensionality reduction similar to PCA,
but allows one to represent both cells and genes in the same space. 

```{r cacomp}
# CA on logcounts
logcounts <- as.matrix(GetAssayData(pbmc, slot = "data"))
ca <- cacomp(obj = logcounts,
             python = TRUE)

# Above is equivalent to:
ca <- cacomp(obj = pbmc,
             assay = "RNA",
             slot = "data",
             python = TRUE)

```

The function `cacomp` takes as input  any non-negative count matrix, be it single-cell, bulk RNA-seq or other data. If we would like to perform feature selection before CA, we can set the argument `top` to the desired number of genes to keep based on the variance of the chi-square components matrix. 
For ease of use it additionally accepts `r CRANpkg("Seurat")` and `r Biocpkg("SingleCellExperiment")` objects, however for these we have to additionally have to specify via the `assay` and/or `slot` (for Seurat) parameter from where to extract the data. The output of `cacomp` is an object of class `cacomp`:
```{r print_cacomp}
ca
```

As can be seen in the summarized output, by default both the standard and principal coordinates in CA space are calculated.
Once we calculate the necessary coordinates for the APL, these will also be shown in the output. Slots are accessed similarly to other S4 classes:
```{r std_coords}
ca@std_coords_cols[1:5,1:5]
```

In the case for both `r CRANpkg("Seurat")` and `r Biocpkg("SingleCellExperiment")` objects, we can alternatively set `return_input = TRUE` to get the input object back, with the CA results stored in the appropriate slot for dimensional reductions. This also enables us to use the plotting functions that come with these packages:

```{r ca_pbmc}
pbmc <- cacomp(obj = pbmc,
               assay = "RNA",
               slot = "data",
               return_input = TRUE,
               python = TRUE)

DimPlot(pbmc, reduction = "CA", label = FALSE, pt.size = 0.5)

```

## Visualization of CA
In order to visualize both cells and genes in a so-called biplot, we however need a `cacomp` object. We can convert to a `cacomp` object from `r CRANpkg("Seurat")` or `r Biocpkg("SingleCellExperiment")` objects that have CA results stored with the function `as.cacomp()`.

```{r biplot, fig.wide = TRUE}
ca <- as.cacomp(pbmc)

platelets <- which(pbmc$cell_type == "8 - Platelet")

# for a static plot
# ca_biplot(ca, type = "ggplot", col_labels = platelets)

# interactive plot
ca_biplot(ca, col_labels = platelets)

```
In order to explore more than 2 dimensions at once, a 3D plot can often be helpful:

```{r 3dplot, fig.wide = TRUE}
ca_3Dplot(ca, col_labels = platelets)

```

The above plots lets us interactively explore the CA space. If a static version is preferred we can obtain it by setting `type = "ggplot"`. This option is available for all plotting functions with interactive plots (with the exception of the 3D plot).

## Choosing the dimensionality
Often we want to only keep as many dimensions of our data as is necessary in order to denoise the data and speed up downstream analysis. For PCA, this is traditionally done by identifying the "bend" in the scree plot of the exaplained variability. For CA we instead look at the inertia that is explained by the data:

```{r scree_plot}
pick_dims(ca,
          method = "scree_plot") +
  xlim(c(0,75))
```

In the scree plot above we can see that the first dimension only explains ~0.75% of the inertia in the data and the the bend in the scree plot is at roughly 10 dimensions. The first dimensions however explain only a small part of the total inertia. 
Instead of relying only on the scree plot, we can use the "elbow rule", a formalized version of the above method. We compare the inertia that is explained by randomizing the data and choose the cutoff as all the dimensions that explain more inertia than the permuted data.

```{r pick_dims, results = "hide"}
pd <- pick_dims(ca,
          mat = GetAssayData(pbmc, slot = "data"),
          method = "elbow_rule",
          reps = 5, 
          python = TRUE)
```
```{r show_dims, message=FALSE}
pd
```
This gives us a much higher number of dimensions. 
```{r expl_inert}
# Calculates explained inertia of each dimension
expl_inertia <- (ca@D^2/sum(ca@D^2))*100

# Sum of expl. intertia kept by elbow method
sum(expl_inertia[seq_len(pd)])
```

In total, the elbow method would suggest to keep `r pd` dimensions that explain `r round(sum(expl_inertia[seq_len(pd)]),2)`% of the total inertia in the data.

Other methods to determine the dimensionality are also implemented:
* "avg_inertia" calculates the number of dimensions in which the inertia is above the average inertia. 
* "maj_inertia" calculates the number of dimensions in which cumulatively explain up to 80% of the total inertia.

Finally, we can subset the data to the desired number of dimensions:
```{r subset_dims}
ca <- subset_dims(ca, dims = pd)
```


# Association Plots

In many scRNA-seq experiments we are interested in which genes define a specific cluster. We can easily visualize these genes in an Association plot. An Association plot allows us to effectivily visualize the relationship between a group of cells (a condition) and genes. Here we chose the platelet cells as our group of interest:
```{r apl_platelets}
# Calculate APL coordinates
ca <- apl_coords(ca, group = platelets)
```

Then we can the results trough the `apl` function. The x-axis can be interpreted as the average coordinates of the platelets. APL allows us to plot the relationship of cells and genes to the chosen condition in 2D, irrespective of the dimensionality of the data.
```{r apl_platelets_plot, fig.wide = TRUE}
# Plot APL
apl(ca,
    row_labs = TRUE,
    rows_idx = c("ITGA2B", "PF4", "GP1BA", "TUBB1"), #platelet marker genes
    type = "ggplot")
```
Cells part of the chosen condition are colored in red, all others in green and genes are blue circles.

The further right a gene is located the more highly associated it is with the chosen condition. Additionally, the lower the y-axis value, the more specific it is to the group. As we can see in the plot, typical Platelet marker genes are located to the right of the plot. If we are interested in not only the genes associated with a condition, but also other cells, we can additionally plot the cells in the same plot through the argument `show_cols = TRUE`. This way we can identify other clusters that are highly similar or cells that should be merged with the chose condition.

The APL Score encompasses both of the above mentioned aspects of condition specificity and allows us to rank genes by how relevant they are for a condition/cluster:

```{r apl_score, results = "hide"}
# Calculates APL score
# For the calculation the input matrix is also required.
ca <- apl_score(ca,
                mat = as.matrix(GetAssayData(object = pbmc, slot = "data")),
                reps = 5,
                python = TRUE)
```
```{r apl_plot_platelets, fig.wide = TRUE}
apl(ca,
    show_score = TRUE,
    show_cols = FALSE,
    type = "plotly")

```

By default, only genes that have a score larger than 0 are colored as these tend to be genes of interest.
However, for some datasets this cutoff is not suitable and can easily be change through the `score_cutoff` argument to `apl()`.

We can also have a look directly at the scores:
```{r print_score}
head(ca@APL_score)
```

In the top 6 most highly ranked genes we can find typical platelet markers such as GP9 (Glycoprotein IX Platelet) or ITGA2B (Integrin Subunit Alpha 2b).
We can also plot these genes through functions provided in `r CRANpkg("Seurat")`:

```{r seurat_apl, fig.wide = TRUE}
VlnPlot(pbmc, features = head(ca@APL_score$Rowname,3))
FeaturePlot(pbmc, features = head(ca@APL_score$Rowname,3))
```

As expected, the 3 most highly scored genes are extremely specific to the platelet cluster.

## APL and Gene Enrichment Analysis

Association plot can be used to help annotate a cluster and to identify a cluster. 
Lets have a closer look at cluster 3. We first calculate the the APL coordinates for both cells and genes and then score the genes.
```{r cluster_three, results="hide"}
# Get indices of cells in cluster 3
c_three <- which(pbmc$seurat_clusters == 3)

# Calculate coordinates and score genes.
ca <- apl_coords(ca, group = c_three)
ca <- apl_score(ca,
                mat = as.matrix(GetAssayData(object = pbmc, slot = "data")),
                reps = 5,
                python = TRUE)
```

We will now use gene overrepresentation analysis as implemented in the package `r Biocpkg("topGO")` on the ranked genes to see if we can annotate the cluster.

```{r topGO, message=FALSE}
enr <- apl_topGO(ca,
          ontology = "BP",
          organism = "hs",
          score_cutoff = 1)
```


```{r topGO_plot, message=FALSE}
plot_enrichment(enr)
```

The gene enrichment clearly suggests that cells in cluster 3 are B cells. We can further look at a number of typical B cell marker genes in the APL.

```{r bcell_apl, fig.wide = TRUE}

bcell_markers <- c("CD19", "MS4A1")

apl(ca,
    show_score = TRUE,
    col_labs = FALSE,
    row_labs = TRUE,
    show_cols = FALSE,
    score_cutoff = 1,
    rows_idx = bcell_markers,
    type = "ggplot")

```

All of these B cell marker genes are clearly highly associated with cluster 3, which confirms the correct annotation we did in the preprocessing of the data.

## APL wrapper

The package also includes a wrapper function `runAPL()` that automates most of the steps above for ease of use. 
```{r}
runAPL(pbmc,
       assay = "RNA",
       slot = "data",
       group = which(pbmc$seurat_clusters == 3),
       score = TRUE,
       dims = 220,
       show_cols = FALSE)
```



# Session info {.unnumbered}

```{r sessionInfo, echo=FALSE}
sessionInfo()
```
