<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Find rows most highly associated with a condition — apl_score • APL</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Find rows most highly associated with a condition — apl_score"><meta name="description" content="Ranks rows by a calculated score which balances the association of the row
with the condition and how associated it is with other conditions."><meta property="og:description" content="Ranks rows by a calculated score which balances the association of the row
with the condition and how associated it is with other conditions."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">APL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.10.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/APL.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Find rows most highly associated with a condition</h1>

      <div class="d-none name"><code>apl_score.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Ranks rows by a calculated score which balances the association of the row
with the condition and how associated it is with other conditions.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">apl_score</span><span class="op">(</span></span>
<span>  <span class="va">caobj</span>,</span>
<span>  mat <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  dims <span class="op">=</span> <span class="va">caobj</span><span class="op">@</span><span class="va">dims</span>,</span>
<span>  group <span class="op">=</span> <span class="va">caobj</span><span class="op">@</span><span class="va">group</span>,</span>
<span>  reps <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  quant <span class="op">=</span> <span class="fl">0.99</span>,</span>
<span>  python <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  store_perm <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"permutation"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-caobj">caobj<a class="anchor" aria-label="anchor" href="#arg-caobj"></a></dt>
<dd><p>A "cacomp" object with principal row coordinates and
standardized column coordinates calculated.</p></dd>


<dt id="arg-mat">mat<a class="anchor" aria-label="anchor" href="#arg-mat"></a></dt>
<dd><p>A numeric matrix. For sequencing a count matrix, gene expression
values with genes in rows and samples/cells in columns.
Should contain row and column names.</p></dd>


<dt id="arg-dims">dims<a class="anchor" aria-label="anchor" href="#arg-dims"></a></dt>
<dd><p>Integer. Number of CA dimensions to retain. Needs to be the same
as in caobj!</p></dd>


<dt id="arg-group">group<a class="anchor" aria-label="anchor" href="#arg-group"></a></dt>
<dd><p>Vector of indices of the columns to calculate centroid/x-axis
direction.</p></dd>


<dt id="arg-reps">reps<a class="anchor" aria-label="anchor" href="#arg-reps"></a></dt>
<dd><p>Integer. Number of permutations to perform.</p></dd>


<dt id="arg-quant">quant<a class="anchor" aria-label="anchor" href="#arg-quant"></a></dt>
<dd><p>Numeric. Single number between 0 and 1 indicating the quantile
used to calculate the cutoff. Default 0.99.</p></dd>


<dt id="arg-python">python<a class="anchor" aria-label="anchor" href="#arg-python"></a></dt>
<dd><p>DEPRACTED. A logical value indicating whether to use singular-value
decomposition from the python package torch.</p></dd>


<dt id="arg-store-perm">store_perm<a class="anchor" aria-label="anchor" href="#arg-store-perm"></a></dt>
<dd><p>Logical. Whether permuted data should be stored in the CA
object.
This implementation dramatically speeds up computation compared to `svd()`
in R.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Method to calculate the cutoff. Either "random" for random
direction method or "permutation" for the permutation method.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns the input "cacomp" object with "APL_score" component added.
APL_score contains a data frame with ranked rows, their score and their
original row number.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The score is calculated by permuting the values of each row to determine the
cutoff angle of the 99<!-- % quantile. -->
$$S_{alpha}(x,y)=x-\frac{y}{\tan\alpha}$$
By default the permutation is repeated 10 times (for random direction min.
300 repetition is recommended!), but for very large matrices
this can be reduced.
The method "permutation" permutes the columns in each row and calculates
AP-coordinates for each such permutation. The cutoff is then taken by the
quantile specified by "quan". The "random" method in contrast calculates
AP-coordinates for the original data, but by looking into random directions.</p>
<p>If store_perm is TRUE the permuted data is stored in the cacomp object and
can be used for future scoring.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Association Plots: Visualizing associations in high-dimensional
correspondence analysis biplots <br>
Elzbieta Gralinska, Martin Vingron <br>
bioRxiv 2020.10.23.352096; doi: https://doi.org/10.1101/2020.10.23.352096</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simulate counts</span></span></span>
<span class="r-in"><span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span>, lambda <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">}</span>,</span></span>
<span class="r-in"><span>               x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run correspondence analysis.</span></span></span>
<span class="r-in"><span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="cacomp.html">cacomp</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cnts</span>, princ_coords <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Parameter top is &gt;nrow(obj) and therefore ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dimensions specified. Setting dimensions to: 9</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Calculate APL coordinates:</span></span></span>
<span class="r-in"><span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="apl_coords.html">apl_coords</a></span><span class="op">(</span><span class="va">ca</span>, group <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Rank genes by S-alpha score</span></span></span>
<span class="r-in"><span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">apl_score</span><span class="op">(</span><span class="va">ca</span>, mat <span class="op">=</span> <span class="va">cnts</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |=======                                                               |  10%  |                                                                              |==============                                                        |  20%  |                                                                              |=====================                                                 |  30%  |                                                                              |============================                                          |  40%  |                                                                              |===================================                                   |  50%  |                                                                              |==========================================                            |  60%  |                                                                              |=================================================                     |  70%  |                                                                              |========================================================              |  80%  |                                                                              |===============================================================       |  90%  |                                                                              |======================================================================| 100%</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Clemens Kohl, Elzbieta Gralinska, Martin Vingron.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

