<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Correspondance Analysis — cacomp • APL</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.4.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.4.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Correspondance Analysis — cacomp"><meta name="description" content="`cacomp` performs correspondence analysis on a matrix or
Seurat/SingleCellExperiment object and returns the transformed data.
`cacomp.seurat` performs correspondence analysis on an assay from a Seurat
container and stores the standardized coordinates of the columns (= cells)
and the principal coordinates of the rows (= genes) as a DimReduc Object in
the Seurat container.
`cacomp.SingleCellExperiment` performs correspondence analysis on an assay
from a SingleCellExperiment and stores the standardized coordinates
 of the columns (= cells) and the principal coordinates of the rows
 (= genes) as a matrix in the SingleCellExperiment container."><meta property="og:description" content="`cacomp` performs correspondence analysis on a matrix or
Seurat/SingleCellExperiment object and returns the transformed data.
`cacomp.seurat` performs correspondence analysis on an assay from a Seurat
container and stores the standardized coordinates of the columns (= cells)
and the principal coordinates of the rows (= genes) as a DimReduc Object in
the Seurat container.
`cacomp.SingleCellExperiment` performs correspondence analysis on an assay
from a SingleCellExperiment and stores the standardized coordinates
 of the columns (= cells) and the principal coordinates of the rows
 (= genes) as a matrix in the SingleCellExperiment container."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">APL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.10.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/APL.html">Get started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Correspondance Analysis</h1>

      <div class="d-none name"><code>cacomp.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>`cacomp` performs correspondence analysis on a matrix or
Seurat/SingleCellExperiment object and returns the transformed data.</p>
<p>`cacomp.seurat` performs correspondence analysis on an assay from a Seurat
container and stores the standardized coordinates of the columns (= cells)
and the principal coordinates of the rows (= genes) as a DimReduc Object in
the Seurat container.</p>
<p>`cacomp.SingleCellExperiment` performs correspondence analysis on an assay
from a SingleCellExperiment and stores the standardized coordinates
 of the columns (= cells) and the principal coordinates of the rows
 (= genes) as a matrix in the SingleCellExperiment container.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">cacomp</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  coords <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  princ_coords <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  python <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  inertia <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  residuals <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'matrix'</span></span>
<span><span class="fu">cacomp</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  coords <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  princ_coords <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  python <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  inertia <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  residuals <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'dgCMatrix'</span></span>
<span><span class="fu">cacomp</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  coords <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  princ_coords <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  python <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  inertia <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  residuals <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'Seurat'</span></span>
<span><span class="fu">cacomp</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  coords <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  princ_coords <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  python <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  inertia <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  residuals <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  assay <span class="op">=</span> <span class="fu">SeuratObject</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span>,</span>
<span>  slot <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  return_input <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># S4 method for class 'SingleCellExperiment'</span></span>
<span><span class="fu">cacomp</span><span class="op">(</span></span>
<span>  <span class="va">obj</span>,</span>
<span>  coords <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  princ_coords <span class="op">=</span> <span class="fl">3</span>,</span>
<span>  python <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  dims <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>  inertia <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  residuals <span class="op">=</span> <span class="st">"pearson"</span>,</span>
<span>  cutoff <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  clip <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">...</span>,</span>
<span>  assay <span class="op">=</span> <span class="st">"counts"</span>,</span>
<span>  return_input <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-obj">obj<a class="anchor" aria-label="anchor" href="#arg-obj"></a></dt>
<dd><p>A numeric matrix or Seurat/SingleCellExperiment object.
For sequencing a count matrix, gene expression values with genes in rows
and samples/cells in columns.
Should contain row and column names.</p></dd>


<dt id="arg-coords">coords<a class="anchor" aria-label="anchor" href="#arg-coords"></a></dt>
<dd><p>Logical. Indicates whether CA standard coordinates should be
calculated.</p></dd>


<dt id="arg-princ-coords">princ_coords<a class="anchor" aria-label="anchor" href="#arg-princ-coords"></a></dt>
<dd><p>Integer. Number indicating whether principal
coordinates should be calculated for the rows (=1), columns (=2),
both (=3) or none (=0).</p></dd>


<dt id="arg-python">python<a class="anchor" aria-label="anchor" href="#arg-python"></a></dt>
<dd><p>DEPRACTED. A logical value indicating whether to use singular-value
decomposition from the python package torch.
This implementation dramatically speeds up computation compared to `svd()`
in R when calculating the full SVD. This parameter only works when dims==NULL
or dims==rank(mat), where caculating a full SVD is demanded.</p></dd>


<dt id="arg-dims">dims<a class="anchor" aria-label="anchor" href="#arg-dims"></a></dt>
<dd><p>Integer. Number of CA dimensions to retain. If NULL:
(0.2 * min(nrow(A), ncol(A)) - 1 ).</p></dd>


<dt id="arg-top">top<a class="anchor" aria-label="anchor" href="#arg-top"></a></dt>
<dd><p>Integer. Number of most variable rows to retain.
Set NULL to keep all.</p></dd>


<dt id="arg-inertia">inertia<a class="anchor" aria-label="anchor" href="#arg-inertia"></a></dt>
<dd><p>Logical. Whether total, row and column inertias should be
calculated and returned.</p></dd>


<dt id="arg-rm-zeros">rm_zeros<a class="anchor" aria-label="anchor" href="#arg-rm-zeros"></a></dt>
<dd><p>Logical. Whether rows &amp; cols containing only 0s should be
removed. Keeping zero only rows/cols might lead to unexpected results.</p></dd>


<dt id="arg-residuals">residuals<a class="anchor" aria-label="anchor" href="#arg-residuals"></a></dt>
<dd><p>character string. Specifies which kind of residuals should
be calculated. Can be "pearson" (default), "freemantukey" or "NB" for
negative-binomial.</p></dd>


<dt id="arg-cutoff">cutoff<a class="anchor" aria-label="anchor" href="#arg-cutoff"></a></dt>
<dd><p>numeric. Residuals that are larger than cutoff or lower than
-cutoff are clipped to cutoff.</p></dd>


<dt id="arg-clip">clip<a class="anchor" aria-label="anchor" href="#arg-clip"></a></dt>
<dd><p>logical. Whether residuals should be clipped if they are
higher/lower than a specified cutoff</p></dd>


<dt id="arg--">...<a class="anchor" aria-label="anchor" href="#arg--"></a></dt>
<dd><p>Other parameters</p></dd>


<dt id="arg-assay">assay<a class="anchor" aria-label="anchor" href="#arg-assay"></a></dt>
<dd><p>Character. The assay from which extract the count matrix for
SVD, e.g. "RNA" for Seurat objects or "counts"/"logcounts" for
SingleCellExperiments.</p></dd>


<dt id="arg-slot">slot<a class="anchor" aria-label="anchor" href="#arg-slot"></a></dt>
<dd><p>character. The slot of the Seurat assay. Default "counts".</p></dd>


<dt id="arg-return-input">return_input<a class="anchor" aria-label="anchor" href="#arg-return-input"></a></dt>
<dd><p>Logical. If TRUE returns the input
(SingleCellExperiment/Seurat object) with the CA results saved in the
reducedDim/DimReduc slot "CA".
 Otherwise returns a "cacomp". Default FALSE.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>Returns a named list of class "cacomp" with components
U, V and D: The results from the SVD.
row_masses and col_masses: Row and columns masses.
top_rows: How many of the most variable rows were retained for the analysis.
tot_inertia, row_inertia and col_inertia: Only if inertia = TRUE.
Total, row and column inertia respectively.</p>
<p>If return_imput = TRUE with Seurat container: Returns input obj of class
"Seurat" with a new Dimensional Reduction Object named "CA".
Standard coordinates of the cells are saved as embeddings,
the principal coordinates of the genes as loadings and
the singular values (= square root of principal intertias/eigenvalues)
are stored as stdev.
To recompute a regular "cacomp" object without rerunning cacomp use
`as.cacomp()`.</p>
<p>If return_input =TRUE for SingleCellExperiment input returns a
SingleCellExperiment object with a matrix of standardized coordinates of
the columns in
reducedDim(obj, "CA"). Additionally, the matrix contains the following
attributes:
"prin_coords_rows": Principal coordinates of the rows.
"singval": Singular values. For the explained inertia of each principal
axis calculate singval^2.
"percInertia": Percent explained inertia of each principal axis.
To recompute a regular "cacomp" object from a SingleCellExperiment without
rerunning cacomp use `as.cacomp()`.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The calculation is performed according to the work of Michael Greenacre.
Singular value decomposition can be performed either with the base R
function `svd` or preferably by the faster pytorch implementation
(python = TRUE). When working with large matrices, CA coordinates and
principal coordinates should only be computed when needed to save
computational time.</p>
    </div>
    <div class="section level2">
    <h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a></h2>
    <p>Greenacre, M. Correspondence Analysis in Practice, Third Edition, 2017.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Simulate scRNAseq data.</span></span></span>
<span class="r-in"><span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>cell_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">5</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   cell_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                   cell_3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run correspondence analysis.</span></span></span>
<span class="r-in"><span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cnts</span>, princ_coords <span class="op">=</span> <span class="fl">3</span>, top <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dimensions specified. Setting dimensions to: 2</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Please consider setting the dimensions to a lower value to speed up the calculation.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Recommended dimensionality: &lt;&lt; min(nrows, ncols) * 0.2</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">###########</span></span></span>
<span class="r-in"><span><span class="co"># Seurat  #</span></span></span>
<span class="r-in"><span><span class="co">###########</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.github.io/seurat-object/" class="external-link">SeuratObject</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simulate counts</span></span></span>
<span class="r-in"><span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span>, lambda <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">}</span>,</span></span>
<span class="r-in"><span>                     x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create Seurat object</span></span></span>
<span class="r-in"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">cnts</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Feature names cannot have underscores ('_'), replacing with dashes ('-')</span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Data is of class matrix. Coercing to dgCMatrix.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run CA and save in dim. reduction slot</span></span></span>
<span class="r-in"><span><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">seu</span>, return_input <span class="op">=</span> <span class="cn">TRUE</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Parameter top is &gt;nrow(obj) and therefore ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dimensions specified. Setting dimensions to: 9</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Run CA and return cacomp object</span></span></span>
<span class="r-in"><span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">seu</span>, return_input <span class="op">=</span> <span class="cn">FALSE</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Parameter top is &gt;nrow(obj) and therefore ignored.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dimensions specified. Setting dimensions to: 9</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">########################</span></span></span>
<span class="r-in"><span><span class="co"># SingleCellExperiment #</span></span></span>
<span class="r-in"><span><span class="co">########################</span></span></span>
<span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Simulate counts</span></span></span>
<span class="r-in"><span><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span>, lambda <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">}</span>,</span></span>
<span class="r-in"><span>               x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">logcnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">cnts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Create SingleCellExperiment object</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span>assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">cnts</span>, logcounts<span class="op">=</span><span class="va">logcnts</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run CA and save in dim. reduction slot.</span></span></span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">sce</span>, return_input <span class="op">=</span> <span class="cn">TRUE</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="co"># on counts</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dimensions specified. Setting dimensions to: 9</span>
<span class="r-in"><span><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">sce</span>, return_input <span class="op">=</span> <span class="cn">TRUE</span>, assay <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span> <span class="co"># on logcounts</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dimensions specified. Setting dimensions to: 9</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># run CA and return cacomp object.</span></span></span>
<span class="r-in"><span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">sce</span>, return_input <span class="op">=</span> <span class="cn">FALSE</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No dimensions specified. Setting dimensions to: 9</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Clemens Kohl, Elzbieta Gralinska, Martin Vingron.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer></div>





  </body></html>

