<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Analyzing data with APL • APL</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing data with APL">
<meta property="og:description" content="APL">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">APL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/APL.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="APL_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Analyzing data with APL</h1>
                        <h4 data-toc-skip class="author">Elzbieta Gralinska</h4>
            <address class="author_afil">
      Max Planck Institute for Molecular Genetics, Berlin, Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:gralinska@molgen.mpg.de" class="email">gralinska@molgen.mpg.de</a>
      </address>
                              <h4 data-toc-skip class="author">Clemens Kohl</h4>
            <address class="author_afil">
      Max Planck Institute for Molecular Genetics, Berlin, Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:kohl@molgen.mpg.de" class="email">kohl@molgen.mpg.de</a>
      </address>
                              <h4 data-toc-skip class="author">Martin Vingron</h4>
            <address class="author_afil">
      Max Planck Institute for Molecular Genetics, Berlin, Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:vingron@molgen.mpg.de" class="email">vingron@molgen.mpg.de</a>
      </address>
                  
      
      
      <div class="hidden name"><code>APL.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This package performs correspondence analysis (CA) and allows to identify cluster-specific genes using Association Plots (AP). Additionally, APL computes the cluster-specificity scores for all genes which allows to rank the genes by their specificity for a selected cell cluster of interest.</p>
    </div>
    
<div class="section level1">
<h1 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h1>
<p>“APL” is a package developed for computation of Association Plots, a method for visualization and analysis of single cell transcriptomics data. The main focus of “APL” is the identification of genes characteristic for individual clusters of cells from input data.</p>
<p>When working with <em>APL</em> package please cite:</p>
<pre><code>Association Plots: Visualizing associations in high-dimensional correspondence analysis biplots
Elzbieta Gralinska, Martin Vingron
bioRxiv 2020.10.23.352096; doi: https://doi.org/10.1101/2020.10.23.352096</code></pre>
<p>A citation can also be obtained in R by running <code>citation("APL")</code>. For a mathematical description of the method, please refer to the manuscript.</p>
</div>
<div class="section level1">
<h1 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h1>
<p>To install the <em>APL</em> from Bioconductor, run:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span>
    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"APL"</span><span class="op">)</span></code></pre></div>
<p>Alternatively the package can also be installed from GitHub:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span>
<span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"VingronLab/APL"</span><span class="op">)</span></code></pre></div>
<p>To additionally build the package vignette, run instead</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://devtools.r-lib.org/reference/remote-reexports.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"VingronLab/APL"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Building the vignette will however take considerable time.</p>
<div class="section level2">
<h2 id="pytorch-installation">pytorch installation<a class="anchor" aria-label="anchor" href="#pytorch-installation"></a>
</h2>
<p>In order to decrease the computation time of the singular value decomposition (SVD), we highly recommend the installation of <code>pytorch</code>. More information on the <code>pytorch</code> installation is given below. Instead of installing <code>pytorch</code>, users can also opt to use the R native SVD. For this, please use the argument <code>python = FALSE</code> wherever applicable in this vignette.</p>
<div class="section level3">
<h3 id="install-pytorch-with-reticulate">Install pytorch with reticulate<a class="anchor" aria-label="anchor" href="#install-pytorch-with-reticulate"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/reticulate" class="external-link">reticulate</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/install_miniconda.html" class="external-link">install_miniconda</a></span><span class="op">(</span><span class="op">)</span> 
<span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html" class="external-link">conda_install</a></span><span class="op">(</span>envname <span class="op">=</span> <span class="st">"r-reticulate"</span>, packages <span class="op">=</span> <span class="st">"numpy"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/pkg/reticulate/man/conda-tools.html" class="external-link">conda_install</a></span><span class="op">(</span>envname <span class="op">=</span> <span class="st">"r-reticulate"</span>, packages <span class="op">=</span> <span class="st">"pytorch"</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level3">
<h3 id="manually-install-pytorch-with-conda">Manually install pytorch with conda<a class="anchor" aria-label="anchor" href="#manually-install-pytorch-with-conda"></a>
</h3>
<p>To install <code>pytorch</code> please download the appropriate Miniconda installer for your system from <a href="https://docs.conda.io/en/latest/miniconda.html" class="external-link">the conda website</a>. Follow the installation instructions on their website and make sure the R package <code>reticulate</code> is also installed before proceeding. Once installed, list all available conda environments via <br><code>conda info --envs</code> <br> One of the environments should have <code>r-reticulate</code> in its name. Depending on where you installed it and your system, the exact path might be different. Activate the environment and install <code>pytorch</code> into it.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="ex">conda</span> activate ~/.local/share/r-miniconda/envs/r-reticulate # change path accordingly.</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a><span class="ex">conda</span> install numpy</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a><span class="ex">conda</span> install pytorch</span></code></pre></div>
</div>
</div>
</div>
<div class="section level1">
<h1 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h1>
<div class="section level2">
<h2 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h2>
<p>In this vignette we will use the 3k Peripheral Blood Mononuclear Cell (PBMC) data from 10x Genomics as an example. To obtain the data necessary to follow the vignette we use the Bioconductor package <em><a href="https://bioconductor.org/packages/3.12/TENxPBMCData" class="external-link">TENxPBMCData</a></em>.</p>
<p>Besides the package <em>APL</em> we will use the single-cell RNA-seq analysis suite <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em> (V. 4.0.4) to preprocess the data, but the preprocessing could equally be done with <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em> and <em><a href="https://bioconductor.org/packages/3.12/scater" class="external-link">scater</a></em>\<em><a href="https://bioconductor.org/packages/3.12/scran" class="external-link">scran</a></em>. For the preprocessing we follow the <a href="https://satijalab.org/seurat/articles/pbmc3k_tutorial.html" class="external-link">Guided Clustering Tutorial</a> from the Seurat vignette. For details about the data preprocessing please refer to their website.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">APL</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">TENxPBMCData</span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="loading-the-data">Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h2>
<p>We start with the loading and preprocessing of the 3k PBMC data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TENxPBMCData/man/TENxPBMCData.html" class="external-link">TENxPBMCData</a></span><span class="op">(</span>dataset <span class="op">=</span> <span class="st">"pbmc3k"</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">rowData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Symbol_TENx</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">$</span><span class="va">Barcode</span>


<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">counts</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span>,
                          assay <span class="op">=</span> <span class="st">"RNA"</span>,
                          project <span class="op">=</span> <span class="st">"pbmc3k"</span>,
                          min.cells <span class="op">=</span> <span class="fl">3</span>,
                          min.features <span class="op">=</span> <span class="fl">200</span>,
                          meta.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">colData</a></span><span class="op">(</span><span class="va">sce</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning: Non-unique features (rownames) present in the input matrix, making</span>
<span class="co">#&gt; unique</span>
<span class="co">#&gt; Warning: The following arguments are not used: row.names</span>
<span class="co">#&gt; Warning: Feature names cannot have underscores ('_'), replacing with dashes</span>
<span class="co">#&gt; ('-')</span>

<span class="va">pbmc</span><span class="op">[[</span><span class="st">"percent.mt"</span><span class="op">]</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/PercentageFeatureSet.html" class="external-link">PercentageFeatureSet</a></span><span class="op">(</span><span class="va">pbmc</span>, pattern <span class="op">=</span> <span class="st">"^MT-"</span><span class="op">)</span>

<span class="co"># Filter data</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">pbmc</span>, subset <span class="op">=</span> <span class="va">nFeature_RNA</span> <span class="op">&gt;</span> <span class="fl">200</span> <span class="op">&amp;</span> <span class="va">nFeature_RNA</span> <span class="op">&lt;</span> <span class="fl">2500</span> <span class="op">&amp;</span> <span class="va">percent.mt</span> <span class="op">&lt;</span> <span class="fl">5</span><span class="op">)</span>
<span class="va">no_zeros_rows</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">0</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="va">pbmc</span><span class="op">[</span><span class="va">no_zeros_rows</span>,<span class="op">]</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="normalization-pca-clustering">Normalization, PCA &amp; Clustering<a class="anchor" aria-label="anchor" href="#normalization-pca-clustering"></a>
</h2>
<p>Association Plots from <em>APL</em> should be computed based on the normalized expression data. Therefore, we first normalize the counts from the 3k PBMC data. For now, <em>APL</em> requires the data to be clustered beforehand. That’s why we also do the cell clustering.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Normalization</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/NormalizeData.html" class="external-link">NormalizeData</a></span><span class="op">(</span><span class="va">pbmc</span>,
                      normalization.method <span class="op">=</span> <span class="st">"LogNormalize"</span>,
                      scale.factor <span class="op">=</span> <span class="fl">10000</span>,
                      verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindVariableFeatures.html" class="external-link">FindVariableFeatures</a></span><span class="op">(</span><span class="va">pbmc</span>,
                             <span class="co"># selection.method = "vst",</span>
                             nfeatures <span class="op">=</span> <span class="fl">2000</span>,
                             verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># Scaling</span>
<span class="va">all.genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/ScaleData.html" class="external-link">ScaleData</a></span><span class="op">(</span><span class="va">pbmc</span>,
                  features <span class="op">=</span> <span class="va">all.genes</span>,
                  verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># Run PCA</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunPCA.html" class="external-link">RunPCA</a></span><span class="op">(</span><span class="va">pbmc</span>,
               features <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/VariableFeatures.html" class="external-link">VariableFeatures</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span><span class="op">)</span>,
               verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="co"># Cell clustering</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindNeighbors.html" class="external-link">FindNeighbors</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/FindClusters.html" class="external-link">FindClusters</a></span><span class="op">(</span><span class="va">pbmc</span>, resolution <span class="op">=</span> <span class="fl">0.5</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>

<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/RunUMAP.html" class="external-link">RunUMAP</a></span><span class="op">(</span><span class="va">pbmc</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span>, verbose <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span>
<span class="co">#&gt; Warning: The default method for RunUMAP has changed from calling Python UMAP via reticulate to the R-native UWOT using the cosine metric</span>
<span class="co">#&gt; To use Python UMAP via reticulate, set umap.method to 'umap-learn' and metric to 'correlation'</span>
<span class="co">#&gt; This message will be shown once per session</span>

<span class="va">new.cluster.ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0 - Naive CD4 T"</span>,
                     <span class="st">"1 - CD14+ Mono"</span>, 
                     <span class="st">"2 - Memory CD4 T"</span>, 
                     <span class="st">"3 - B"</span>, 
                     <span class="st">"4 - CD8 T"</span>, 
                     <span class="st">"5 - FCGR3A+ Mono"</span>, 
                     <span class="st">"6 - NK"</span>, 
                     <span class="st">"7 - DC"</span>, 
                     <span class="st">"8 - Platelet"</span><span class="op">)</span>

<span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">new.cluster.ids</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/levels.html" class="external-link">levels</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>
<span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html" class="external-link">RenameIdents</a></span><span class="op">(</span><span class="va">pbmc</span>, <span class="va">new.cluster.ids</span><span class="op">)</span>
<span class="va">pbmc</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/Idents.html" class="external-link">Idents</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>

<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"umap"</span>, label <span class="op">=</span> <span class="cn">FALSE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="APL_files/figure-html/preprocessing-1.png" width="700"></p>
</div>
</div>
<div class="section level1">
<h1 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h1>
<p>The fastest way to compute the Association Plot for a selected cluster of cells from the input data is by using a wrapper function <code><a href="../reference/runAPL.html">runAPL()</a></code>. <code><a href="../reference/runAPL.html">runAPL()</a></code> automates most of the analysis steps for ease of use.</p>
<p>For example, to generate the Association Plot for the B cells (cluster: “3 - B”) we can use the following command:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="kw">runAPL</span>(pbmc,</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>       <span class="dt">assay =</span> <span class="st">"RNA"</span>,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a>       <span class="dt">slot =</span> <span class="st">"data"</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a>       <span class="dt">top =</span> <span class="kw">nrow</span>(pbmc),</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a>       <span class="dt">group =</span> <span class="kw">which</span>(pbmc<span class="op">$</span>seurat_clusters <span class="op">==</span><span class="st"> </span><span class="dv">3</span>),</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a>       <span class="dt">type =</span> <span class="st">"ggplot"</span>)</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true"></a><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true"></a><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true"></a><span class="co">#&gt;  Using 222 dimensions. Subsetting.</span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true"></a><span class="co">#&gt; </span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true"></a><span class="st">  </span><span class="er">|</span><span class="st">                                                                      </span><span class="er">|</span><span class="st">   </span><span class="dv">0</span>%</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true"></a><span class="st">  </span><span class="er">|=======================</span><span class="st">                                               </span><span class="er">|</span><span class="st">  </span><span class="dv">33</span>%</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true"></a><span class="st">  </span><span class="er">|===============================================</span><span class="st">                       </span><span class="er">|</span><span class="st">  </span><span class="dv">67</span>%</span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true"></a>  <span class="op">|</span><span class="st">                                                                            </span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true"></a><span class="st">  </span><span class="er">|======================================================================|</span><span class="st"> </span><span class="dv">100</span>%</span></code></pre></div>
<p><img src="APL_files/figure-html/runAPL-1.png" width="700"> The generated Association Plot is computed based on the log-normalized count matrix. By default <code>runAPL</code> uses the top 5,000 most variable genes in the data, but the data can be subset to any number of genes by changing the value for the argument <code>top</code>. In the above-presented example we keep all genes to ensure that marker genes for smaller clusters are also included. The dimensionality of the CA is determined automatically by the elbow rule descibed below (see <a href="#dim_reduc">here</a>). This default behaviour can be overriden by setting the dimensions manually (parameter <code>dims</code>). The cluster-specificity score ( ) for each gene is also calculated (<code>score = TRUE</code>). In order to better explore the data, <code>type</code> can be set to <code>"plotly"</code> to obtain an interactive plot. <code>runAPL</code> has many arguments to further customize the output and fine tune the calculations. Please refer to the documentation (<code><a href="../reference/runAPL.html">?runAPL</a></code>) for more information. The following sections in this vignette will discuss the choice of dimensionality and the -score.</p>
</div>
<div class="section level1">
<h1 id="step-by-step-way-of-computing-association-plots">Step-by-step way of computing Association Plots<a class="anchor" aria-label="anchor" href="#step-by-step-way-of-computing-association-plots"></a>
</h1>
<p>Alternatively, Association Plots can be computed step-by-step. This allows to adjust the Association Plots to user’s needs. Below we explain each step of the process of generating Association Plots.</p>
<div class="section level2">
<h2 id="correspondence-analysis">Correspondence Analysis<a class="anchor" aria-label="anchor" href="#correspondence-analysis"></a>
</h2>
<p>The first step of Association Plot computations is correspondence analysis (CA). CA is a data dimensionality reduction method similar to PCA, however it allows for a simultaneous embedding of both cells and genes from the input data in the same space. In this example we perform CA on the log-normalized count matrix from the 3k PBMC data.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Computing CA on logcounts</span>
<span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span><span class="op">)</span>
<span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp.html">cacomp</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">logcounts</span>,
             top <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,
             python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="co"># The above is equivalent to:</span>
<span class="co"># ca &lt;- cacomp(obj = pbmc,</span>
<span class="co">#              assay = "RNA",</span>
<span class="co">#              slot = "data",</span>
<span class="co">#              top = nrow(pbmc),</span>
<span class="co">#              python = TRUE)</span></code></pre></div>
<p>The function <code>cacomp</code> accepts as an input any matrix with non-negative entries, be it a single-cell RNA-seq, bulk RNA-seq or other data. For ease of use, <code>cacomp</code> accepts also <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em> and <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em> objects, however for these we additionally have to specify via the <code>assay</code> and/or <code>slot</code> (for Seurat) parameter from where to extract the data. Importantly, in order to ensure the interpretability of the results <code>cacomp</code> (and related functions such as <code>runAPL</code>) requires that the input matrix contains both row and column names.</p>
<p>When performing a feature selection before CA, we can set the argument <code>top</code> to the desired number of genes with the highest variance across cells from the input data to retain for further analysis. By default, only the top 5,000 most variable genes are kept. For this vignette we keep all genes in the data to ensure that also genes from smaller clusters such as the platelets are kept. In many cases however 5,000 genes are a good compromise between computational time and keeping most relevant genes.</p>
<p>The output of <code>cacomp</code> is an object of class <code>cacomp</code>:</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ca</span>
<span class="co">#&gt; cacomp object with 2638 columns, 13713 rows and 2637 dimensions.</span>
<span class="co">#&gt; Calc. standard coord.:  std_coords_rows, std_coords_cols</span>
<span class="co">#&gt; Calc. principal coord.: prin_coords_rows, prin_coords_cols</span>
<span class="co">#&gt; Calc. APL coord.:       </span>
<span class="co">#&gt; Explained inertia:      0.8% Dim1, 0.6% Dim2</span></code></pre></div>
<p>As can be seen in the summarized output, by default both types of coordinates in the CA space (principal and standardized) are calculated. Once the coordinates for the Association Plot are calculated, they will also be shown in the output of <code>cacomp</code>. Slots are accessed through an accessor function:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"std_coords_cols"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span>
<span class="co">#&gt;                         Dim1        Dim2        Dim3        Dim4         Dim5</span>
<span class="co">#&gt; AAACATACAACCAC-1  0.69703119 -0.09113136 -0.41134163  0.32559771 -0.272531983</span>
<span class="co">#&gt; AAACATTGAGCTAC-1  0.07402344  0.02951448  1.40628242 -1.36663142 -0.154395581</span>
<span class="co">#&gt; AAACATTGATCAGC-1  0.60068885 -0.12408014 -0.50940106  0.53809503 -0.001027849</span>
<span class="co">#&gt; AAACCGTGCTTCCG-1 -1.50638399  0.43243122 -0.02952732  0.03923192 -0.486999974</span>
<span class="co">#&gt; AAACCGTGTATGCG-1  0.27353640 -0.02000082 -2.00853340 -2.85132099  0.085657719</span></code></pre></div>
<p>In the case of <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em> and <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em> objects, we can alternatively set <code>return_input = TRUE</code> to get the input object back, with the CA results computed by “APL” and stored in the appropriate slot for dimension reduction. This also allows for using the plotting functions that come with these packages:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp.html">cacomp</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>,
               assay <span class="op">=</span> <span class="st">"RNA"</span>,
               slot <span class="op">=</span> <span class="st">"data"</span>,
               top <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span>,
               return_input <span class="op">=</span> <span class="cn">TRUE</span>,
               python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>

<span class="fu"><a href="https://satijalab.org/seurat/reference/DimPlot.html" class="external-link">DimPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, reduction <span class="op">=</span> <span class="st">"CA"</span>, label <span class="op">=</span> <span class="cn">FALSE</span>, pt.size <span class="op">=</span> <span class="fl">0.5</span><span class="op">)</span></code></pre></div>
<p><img src="APL_files/figure-html/ca_pbmc-1.png" width="700"></p>
<p>However, some functions such as apl_coords() require information that cannot be stored in the single-cell container objects. It is therefore often easier to work with a <code>cacomp</code> object instead. We can convert <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em> or <em><a href="https://bioconductor.org/packages/3.12/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em> objects which have CA results stored to a <code>cacomp</code> object using the function <code><a href="../reference/as.cacomp.html">as.cacomp()</a></code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Converting the object pbmc to cacomp</span>
<span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.cacomp.html">as.cacomp</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="dim_reduc">Reducing the number of CA dimensions<a class="anchor" aria-label="anchor" href="#dim_reduc"></a>
</h2>
<p>When working with high-dimensional data, after singular value decomposition there will be often many dimensions which will be representing the noise in the data. That’s why generating Association Plots should be preceded by data dimension reduction step.</p>
<p>The number of dimensions to retain can be computed using the function <code>pick_dims</code>. This function offers three standard methods which we implemented:</p>
<ul>
<li><p>elbow rule (<code>method = "elbow_rule"</code>) - the number of dimensions to retain is calculated based on scree plots generated for randomized data, and corresponds to a point in the plot where the band of randomized singular values enters the band of the original singular values,</p></li>
<li><p>80% rule (<code>method = "maj_inertia"</code>) - only those first dimensions are retained which in total account for &gt;= 80% of total inertia,</p></li>
<li><p>average rule (<code>method = "avg_inertia"</code>) - only those dimensions are retained which account for more inertia than a single dimension on average.</p></li>
</ul>
<p>Additionally, the user can compute a scree plot to choose the number of dimensions by themselves:</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/pick_dims.html">pick_dims</a></span><span class="op">(</span><span class="va">ca</span>,
          method <span class="op">=</span> <span class="st">"scree_plot"</span><span class="op">)</span> <span class="op">+</span>
  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">75</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; Warning: Removed 2562 rows containing missing values (position_stack).</span>
<span class="co">#&gt; Warning: Removed 1 rows containing missing values (geom_col).</span>
<span class="co">#&gt; Warning: Removed 2562 row(s) containing missing values (geom_path).</span></code></pre></div>
<p><img src="APL_files/figure-html/scree_plot-1.png" width="700"></p>
<p>In the scree plot above we can see that the first dimension explains only ~0.75% of the total inertia and we observe the “jump” in the scree plot at roughly 15 dimensions. The first dimensions however explain only a small amount of the total inertia.</p>
<p>Here we compute the number of dimensions using the elbow rule. For demonstration, only five data permutations are computed:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pick_dims.html">pick_dims</a></span><span class="op">(</span><span class="va">ca</span>,
          mat <span class="op">=</span> <span class="fu"><a href="https://satijalab.org/seurat/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span><span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span>,
          method <span class="op">=</span> <span class="st">"elbow_rule"</span>,
          reps <span class="op">=</span> <span class="fl">5</span>, 
          python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pd</span>
<span class="co">#&gt; [1] 223</span></code></pre></div>
<p>In this case the elbow rule leads to a much higher number of dimensions.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Compute the amount of inertia explained by each of the dimensions</span>
<span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"D"</span><span class="op">)</span>
<span class="va">expl_inertia</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">D</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">D</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span>

<span class="co"># Compute the amount of intertia explained </span>
<span class="co"># by the number of dimensions defined by elbow rule</span>
<span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">expl_inertia</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>
<span class="co">#&gt; [1] 18.654</span></code></pre></div>
<p>In this example the elbow rule suggests to keep 223 dimensions that explain 18.65% of the total inertia from the data.</p>
<p>Finally, we can reduce the data dimension to the desired number of dimensions:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subset_dims.html">subset_dims</a></span><span class="op">(</span><span class="va">ca</span>, dims <span class="op">=</span> <span class="va">pd</span><span class="op">)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="association-plots">Association Plots<a class="anchor" aria-label="anchor" href="#association-plots"></a>
</h2>
<p>When working with single-cell transcriptomics data we are often interested in which genes are associated to a cluster of cells. To reveal such genes we can compute an Association Plot for a selected cluster of cells. In the following example we want to generate an Association Plot for the cluster of platelets:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Specifying a cell cluster of interest</span>
<span class="va">platelets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">==</span> <span class="st">"8 - Platelet"</span><span class="op">)</span>

<span class="co"># Calculate Association Plot coordinates for platelets from 3k PBMC data</span>
<span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_coords.html">apl_coords</a></span><span class="op">(</span><span class="va">ca</span>, group <span class="op">=</span> <span class="va">platelets</span><span class="op">)</span></code></pre></div>
<p>After computing the coordinates of genes and cells in the Association Plot we are able to plot the results using the <code>apl</code> function.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Plot APL</span>
<span class="fu"><a href="../reference/apl.html">apl</a></span><span class="op">(</span><span class="va">ca</span>,
    row_labs <span class="op">=</span> <span class="cn">TRUE</span>,
    rows_idx <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ITGA2B"</span>, <span class="st">"PF4"</span>, <span class="st">"GP1BA"</span>, <span class="st">"TUBB1"</span><span class="op">)</span>, <span class="co"># platelet marker genes</span>
    type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span> <span class="co"># type = "plotly" for an interactive plot</span></code></pre></div>
<p><img src="APL_files/figure-html/apl_platelets_plot-1.png" width="700"> In the Association Plot all genes are represented by blue circles. The further to the right a gene is located the more associated it is with the chosen cluster of cells. Additionally, the lower the y-axis value, the more specific it is for the selected cluster. Additionally, it is possible to highlight in the Association Plot any set of genes. In the example above we additionally highlighted four genes (“ITGA2B”, “PF4”, “GP1BA”, “TUBB1”) which are known to be marker genes for platelets. As we can see in the plot, they are located in the right part of the plot, which confirms their specificity for platelets.</p>
<p>By default we plot only the genes in the Association Plot. To also display the cells in the Association Plot, use the argument <code>show_cols = TRUE</code>. This way we can identify other cells which show similar expression profiles to the cells of interest. Cells that belong to the cluster of interest will be colored in red, and all remaining cells will be colored in violet. Furthermore, an interactive plot in which you can hover over genes to see their name can be created by setting <code>type = "plotly"</code>.</p>
</div>
<div class="section level2">
<h2 id="association-plots-with-the-scores">Association Plots with the  scores<a class="anchor" aria-label="anchor" href="#association-plots-with-the-scores"></a>
</h2>
<p>The  score allows us to rank genes by their specificity for a selected cell cluster, and is computed for each gene from the Association Plot separately. The higher the  score of a gene, the more characteristic its expression for the investigated cell cluster. The  scores can be computed using the <code>apl_score</code> function. To show the  scores in the Association Plot use the argument <code>show_score = TRUE</code> in the <code>apl</code> function:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Compute S-alpha score</span>
<span class="co"># For the calculation the input matrix is also required.</span>
<span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_score.html">apl_score</a></span><span class="op">(</span><span class="va">ca</span>,
                mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span><span class="op">)</span>,
                reps <span class="op">=</span> <span class="fl">5</span>,
                python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/apl.html">apl</a></span><span class="op">(</span><span class="va">ca</span>,
    show_score <span class="op">=</span> <span class="cn">TRUE</span>,
    type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span> </code></pre></div>
<p><img src="APL_files/figure-html/apl_plot_platelets-1.png" width="700"></p>
<p>By default, only genes that have a  score larger than 0 are colored as these tend to be genes of interest and we consider them as cluster-specific genes. This cutoff can be easily changed through the <code>score_cutoff</code> argument to <code><a href="../reference/apl.html">apl()</a></code>.</p>
<p>The  scores are stored in the <code>"APL_score"</code> slot and can be accessed as follows:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"APL_score"</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt;               Rowname    Score Row_num Rank</span>
<span class="co">#&gt; AP001189.4 AP001189.4 15.49974    8015    1</span>
<span class="co">#&gt; ITGA2B         ITGA2B 15.24879   11215    2</span>
<span class="co">#&gt; GP9               GP9 15.22077    2854    3</span>
<span class="co">#&gt; LY6G6F         LY6G6F 13.99465    4381    4</span>
<span class="co">#&gt; SEPT5           SEPT5 13.64891   13232    5</span>
<span class="co">#&gt; PVALB           PVALB 13.25364   13384    6</span></code></pre></div>
<p>To see the expression of genes with the highest  scores (or any selected genes) across all cell types from the data we can use the functions provided by <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em>:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"APL_score"</span><span class="op">)</span>
<span class="fu"><a href="https://satijalab.org/seurat/reference/VlnPlot.html" class="external-link">VlnPlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scores</span><span class="op">$</span><span class="va">Rowname</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="APL_files/figure-html/seurat_apl-1.png" width="700"></p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://satijalab.org/seurat/reference/FeaturePlot.html" class="external-link">FeaturePlot</a></span><span class="op">(</span><span class="va">pbmc</span>, features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scores</span><span class="op">$</span><span class="va">Rowname</span>,<span class="fl">3</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<p><img src="APL_files/figure-html/seurat_apl-2.png" width="700"></p>
<p>As expected, the 3 most highly scored genes are over-expressed in the platelet cluster.</p>
</div>
<div class="section level2">
<h2 id="visualization-of-ca">Visualization of CA<a class="anchor" aria-label="anchor" href="#visualization-of-ca"></a>
</h2>
<p>In addition to Association Plots “APL” produces also other forms of the output. For instance, we can use “APL” to generate a two- and three-dimensional correspondence analysis projection of the data. The so-called biplot visualizes both cells and genes from the input data and can be created using the function <code>ca_biplot</code>. Alternatively, a three-dimensional data projection plot can be generated using the function <code>ca_3Dplot</code>. To generate such biplots a <code>cacomp</code> object is required.</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Specifying a cell cluster of interest</span>
<span class="va">platelets</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">cell_type</span> <span class="op">==</span> <span class="st">"8 - Platelet"</span><span class="op">)</span>

<span class="co"># Creating a static plot</span>
<span class="fu"><a href="../reference/ca_biplot.html">ca_biplot</a></span><span class="op">(</span><span class="va">ca</span>, col_labels <span class="op">=</span> <span class="va">platelets</span>, type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span></code></pre></div>
<p><img src="APL_files/figure-html/biplot-1.png" width="700"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="co"># Creating an interactive plot</span>
<span class="co"># ca_biplot(ca, type = "plotly", col_labels = platelets)</span>

<span class="co"># 3D plot</span>
<span class="co"># ca_3Dplot(ca, col_labels = platelets)</span></code></pre></div>
<p>The above described plots give us a quick overview of the first 2 dimensions of the data (more dimensions can be plotted). As shown in the commented-out code, to interactively explore the projection of the data <code>type = "plotly"</code> can be set.</p>
</div>
</div>
<div class="section level1">
<h1 id="apl-and-go-enrichment-analysis">APL and GO enrichment analysis<a class="anchor" aria-label="anchor" href="#apl-and-go-enrichment-analysis"></a>
</h1>
<p>After computing an Association Plot and identifying a set of genes specific for a selected cluster of cells we might be interested in conducting a Gene Ontology (GO) enrichment analysis of the identified gene set. To conduct the GO enrichment analysis of the B-cell specific genes (cluster 3 from the input data) idenitfied using Association Plot we first need to compute the coordinates of the genes in the Association Plot for B cells, as well as the  score for each gene:</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># Get indices of cells in cluster 3 (B cells)</span>
<span class="va">c_three</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">pbmc</span><span class="op">$</span><span class="va">seurat_clusters</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span>

<span class="co"># Calculate Association Plot coordinates of the genes and the \eqn{S_\alpha} scores</span>
<span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_coords.html">apl_coords</a></span><span class="op">(</span><span class="va">ca</span>, group <span class="op">=</span> <span class="va">c_three</span><span class="op">)</span>
<span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_score.html">apl_score</a></span><span class="op">(</span><span class="va">ca</span>,
                mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="fu"><a href="https://satijalab.org/seurat/reference/AssayData.html" class="external-link">GetAssayData</a></span><span class="op">(</span>object <span class="op">=</span> <span class="va">pbmc</span>, slot <span class="op">=</span> <span class="st">"data"</span><span class="op">)</span><span class="op">)</span>,
                reps <span class="op">=</span> <span class="fl">5</span>,
                python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<p>Now we can conduct GO enrichment analysis as implemented in the package <em><a href="https://bioconductor.org/packages/3.12/topGO" class="external-link">topGO</a></em> using the most cluster-specific genes from the Association Plot. By default we use all genes with an  score higher than 0, but the cutoff may have to be adjusted depending on the dataset. In the example below we restrict it to genes with a  score higher than 1 to restrict it to truly significant genes. In case that no  scores were calculated, one can also choose to use the <code>ngenes</code> (by default 1000) genes with the highest x-coordinates by setting <code>use_coords = FALSE</code>.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">enr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_topGO.html">apl_topGO</a></span><span class="op">(</span><span class="va">ca</span>,
                 ontology <span class="op">=</span> <span class="st">"BP"</span>,
                 organism <span class="op">=</span> <span class="st">"hs"</span>,
                 score_cutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></code></pre></div>
<p>The function <code><a href="../reference/plot_enrichment.html">plot_enrichment()</a></code> was implemented to visualize the <code>topGO</code> results in form of a dotplot.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/plot_enrichment.html">plot_enrichment</a></span><span class="op">(</span><span class="va">enr</span><span class="op">)</span></code></pre></div>
<p><img src="APL_files/figure-html/topGO_plot-1.png" width="700"></p>
<p>Additionally, in the Association Plot we can highlight genes of interest identified for instance using the GO enrichment analysis. In the example below we highlight two canonical B cell markers.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R">
<span class="va">bcell_markers</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD19"</span>, <span class="st">"MS4A1"</span><span class="op">)</span>

<span class="fu"><a href="../reference/apl.html">apl</a></span><span class="op">(</span><span class="va">ca</span>,
    show_score <span class="op">=</span> <span class="cn">TRUE</span>,
    col_labs <span class="op">=</span> <span class="cn">FALSE</span>,
    row_labs <span class="op">=</span> <span class="cn">TRUE</span>,
    score_cutoff <span class="op">=</span> <span class="fl">1</span>,
    rows_idx <span class="op">=</span> <span class="va">bcell_markers</span>,
    type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span></code></pre></div>
<p><img src="APL_files/figure-html/bcell_apl-1.png" width="700"></p>
</div>
<div class="section level1 unnumbered">
<h1 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h1>
<pre><code><span class="co">#&gt; R version 4.0.3 (2020-10-10)</span>
<span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span>
<span class="co">#&gt; Running under: Debian GNU/Linux 10 (buster)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Matrix products: default</span>
<span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3</span>
<span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.12.so</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; locale:</span>
<span class="co">#&gt;  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span>
<span class="co">#&gt;  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span>
<span class="co">#&gt;  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span>
<span class="co">#&gt;  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span>
<span class="co">#&gt;  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span>
<span class="co">#&gt; [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; attached base packages:</span>
<span class="co">#&gt; [1] parallel  stats4    stats     graphics  grDevices utils     datasets </span>
<span class="co">#&gt; [8] methods   base     </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; other attached packages:</span>
<span class="co">#&gt;  [1] SparseM_1.81                org.Hs.eg.db_3.12.0        </span>
<span class="co">#&gt;  [3] AnnotationDbi_1.52.0        ggplot2_3.3.3              </span>
<span class="co">#&gt;  [5] TENxPBMCData_1.8.0          HDF5Array_1.18.1           </span>
<span class="co">#&gt;  [7] rhdf5_2.34.0                DelayedArray_0.16.0        </span>
<span class="co">#&gt;  [9] Matrix_1.4-0                SingleCellExperiment_1.12.0</span>
<span class="co">#&gt; [11] SummarizedExperiment_1.20.0 Biobase_2.50.0             </span>
<span class="co">#&gt; [13] GenomicRanges_1.42.0        GenomeInfoDb_1.26.1        </span>
<span class="co">#&gt; [15] IRanges_2.24.0              S4Vectors_0.28.0           </span>
<span class="co">#&gt; [17] BiocGenerics_0.36.0         MatrixGenerics_1.2.0       </span>
<span class="co">#&gt; [19] matrixStats_0.57.0          SeuratObject_4.0.1         </span>
<span class="co">#&gt; [21] Seurat_4.0.1                APL_0.99.0                 </span>
<span class="co">#&gt; [23] BiocStyle_2.18.1           </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; loaded via a namespace (and not attached):</span>
<span class="co">#&gt;   [1] AnnotationHub_2.22.1          BiocFileCache_1.14.0         </span>
<span class="co">#&gt;   [3] systemfonts_1.0.3             plyr_1.8.6                   </span>
<span class="co">#&gt;   [5] igraph_1.2.6                  lazyeval_0.2.2               </span>
<span class="co">#&gt;   [7] splines_4.0.3                 listenv_0.8.0                </span>
<span class="co">#&gt;   [9] scattermore_0.7               digest_0.6.27                </span>
<span class="co">#&gt;  [11] htmltools_0.5.2               GO.db_3.12.1                 </span>
<span class="co">#&gt;  [13] fansi_0.4.1                   magrittr_2.0.1               </span>
<span class="co">#&gt;  [15] memoise_2.0.0                 tensor_1.5                   </span>
<span class="co">#&gt;  [17] cluster_2.1.0                 ROCR_1.0-11                  </span>
<span class="co">#&gt;  [19] globals_0.14.0                pkgdown_2.0.1                </span>
<span class="co">#&gt;  [21] spatstat.sparse_2.0-0         colorspace_2.0-0             </span>
<span class="co">#&gt;  [23] rappdirs_0.3.1                blob_1.2.1                   </span>
<span class="co">#&gt;  [25] ggrepel_0.9.1                 textshaping_0.3.6            </span>
<span class="co">#&gt;  [27] xfun_0.28                     dplyr_1.0.6                  </span>
<span class="co">#&gt;  [29] crayon_1.4.1                  RCurl_1.98-1.2               </span>
<span class="co">#&gt;  [31] jsonlite_1.7.2                org.Mm.eg.db_3.12.0          </span>
<span class="co">#&gt;  [33] graph_1.68.0                  spatstat.data_2.1-0          </span>
<span class="co">#&gt;  [35] survival_3.2-7                zoo_1.8-9                    </span>
<span class="co">#&gt;  [37] glue_1.4.2                    polyclip_1.10-0              </span>
<span class="co">#&gt;  [39] gtable_0.3.0                  zlibbioc_1.36.0              </span>
<span class="co">#&gt;  [41] XVector_0.30.0                leiden_0.3.7                 </span>
<span class="co">#&gt;  [43] Rhdf5lib_1.12.1               future.apply_1.7.0           </span>
<span class="co">#&gt;  [45] abind_1.4-5                   scales_1.1.1                 </span>
<span class="co">#&gt;  [47] DBI_1.1.0                     miniUI_0.1.1.1               </span>
<span class="co">#&gt;  [49] Rcpp_1.0.5                    viridisLite_0.4.0            </span>
<span class="co">#&gt;  [51] xtable_1.8-4                  reticulate_1.20              </span>
<span class="co">#&gt;  [53] spatstat.core_2.1-2           bit_4.0.4                    </span>
<span class="co">#&gt;  [55] htmlwidgets_1.5.3             httr_1.4.2                   </span>
<span class="co">#&gt;  [57] RColorBrewer_1.1-2            ellipsis_0.3.2               </span>
<span class="co">#&gt;  [59] ica_1.0-2                     farver_2.0.3                 </span>
<span class="co">#&gt;  [61] pkgconfig_2.0.3               dbplyr_2.1.1                 </span>
<span class="co">#&gt;  [63] uwot_0.1.10                   deldir_0.2-10                </span>
<span class="co">#&gt;  [65] sass_0.4.0                    utf8_1.1.4                   </span>
<span class="co">#&gt;  [67] labeling_0.4.2                tidyselect_1.1.0             </span>
<span class="co">#&gt;  [69] rlang_0.4.11                  reshape2_1.4.4               </span>
<span class="co">#&gt;  [71] later_1.1.0.1                 BiocVersion_3.12.0           </span>
<span class="co">#&gt;  [73] munsell_0.5.0                 tools_4.0.3                  </span>
<span class="co">#&gt;  [75] cachem_1.0.5                  ExperimentHub_1.16.1         </span>
<span class="co">#&gt;  [77] generics_0.1.0                RSQLite_2.2.1                </span>
<span class="co">#&gt;  [79] ggridges_0.5.3                evaluate_0.14                </span>
<span class="co">#&gt;  [81] stringr_1.4.0                 fastmap_1.1.0                </span>
<span class="co">#&gt;  [83] goftest_1.2-2                 yaml_2.2.1                   </span>
<span class="co">#&gt;  [85] ragg_1.2.1                    knitr_1.36                   </span>
<span class="co">#&gt;  [87] bit64_4.0.5                   fs_1.5.0                     </span>
<span class="co">#&gt;  [89] fitdistrplus_1.1-3            purrr_0.3.4                  </span>
<span class="co">#&gt;  [91] RANN_2.6.1                    nlme_3.1-151                 </span>
<span class="co">#&gt;  [93] pbapply_1.4-3                 future_1.20.1                </span>
<span class="co">#&gt;  [95] mime_0.9                      compiler_4.0.3               </span>
<span class="co">#&gt;  [97] interactiveDisplayBase_1.28.0 curl_4.3                     </span>
<span class="co">#&gt;  [99] plotly_4.9.3                  png_0.1-7                    </span>
<span class="co">#&gt; [101] spatstat.utils_2.1-0          tibble_3.1.2                 </span>
<span class="co">#&gt; [103] bslib_0.3.1                   stringi_1.5.3                </span>
<span class="co">#&gt; [105] highr_0.8                     RSpectra_0.16-0              </span>
<span class="co">#&gt; [107] desc_1.3.0                    lattice_0.20-41              </span>
<span class="co">#&gt; [109] vctrs_0.3.8                   rhdf5filters_1.2.1           </span>
<span class="co">#&gt; [111] pillar_1.6.1                  lifecycle_1.0.0              </span>
<span class="co">#&gt; [113] BiocManager_1.30.15           spatstat.geom_2.1-0          </span>
<span class="co">#&gt; [115] lmtest_0.9-38                 jquerylib_0.1.4              </span>
<span class="co">#&gt; [117] RcppAnnoy_0.0.18              data.table_1.13.2            </span>
<span class="co">#&gt; [119] cowplot_1.1.1                 bitops_1.0-6                 </span>
<span class="co">#&gt; [121] irlba_2.3.3                   httpuv_1.5.4                 </span>
<span class="co">#&gt; [123] patchwork_1.1.1               R6_2.5.0                     </span>
<span class="co">#&gt; [125] bookdown_0.24                 promises_1.1.1               </span>
<span class="co">#&gt; [127] topGO_2.42.0                  KernSmooth_2.23-18           </span>
<span class="co">#&gt; [129] gridExtra_2.3                 parallelly_1.21.0            </span>
<span class="co">#&gt; [131] codetools_0.2-18              MASS_7.3-53                  </span>
<span class="co">#&gt; [133] assertthat_0.2.1              rprojroot_2.0.2              </span>
<span class="co">#&gt; [135] withr_2.4.2                   sctransform_0.3.2            </span>
<span class="co">#&gt; [137] GenomeInfoDbData_1.2.4        mgcv_1.8-33                  </span>
<span class="co">#&gt; [139] rpart_4.1-15                  grid_4.0.3                   </span>
<span class="co">#&gt; [141] tidyr_1.1.3                   rmarkdown_2.11               </span>
<span class="co">#&gt; [143] Rtsne_0.15                    shiny_1.5.0</span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Elzbieta Gralinska, Clemens Kohl, Martin Vingron.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
