<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Find rows most highly associated with a condition — apl_score • APL</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Find rows most highly associated with a condition — apl_score"><meta property="og:description" content="Ranks rows by a calculated score which balances the association of the row 
with the condition and how associated it is with other conditions."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">APL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/APL.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Find rows most highly associated with a condition</h1>
    
    <div class="hidden name"><code>apl_score.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Ranks rows by a calculated score which balances the association of the row 
with the condition and how associated it is with other conditions.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">apl_score</span><span class="op">(</span>
  <span class="va">caobj</span>,
  <span class="va">mat</span>,
  dims <span class="op">=</span> <span class="va">caobj</span><span class="op">@</span><span class="va">dims</span>,
  group <span class="op">=</span> <span class="va">caobj</span><span class="op">@</span><span class="va">group</span>,
  reps <span class="op">=</span> <span class="fl">10</span>,
  quant <span class="op">=</span> <span class="fl">0.99</span>,
  python <span class="op">=</span> <span class="cn">TRUE</span>,
  store_perm <span class="op">=</span> <span class="cn">TRUE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>caobj</dt>
<dd><p>A "cacomp" object with principal row coordinates and 
standardized column coordinates calculated.</p></dd>
<dt>mat</dt>
<dd><p>A numeric matrix. For sequencing a count matrix, gene expression 
values with genes in rows and samples/cells in columns.
Should contain row and column names.</p></dd>
<dt>dims</dt>
<dd><p>Integer. Number of CA dimensions to retain. Needs to be the same 
as in caobj!</p></dd>
<dt>group</dt>
<dd><p>Vector of indices of the columns to calculate centroid/x-axis 
direction.</p></dd>
<dt>reps</dt>
<dd><p>Integer. Number of permutations to perform. Default = 10.</p></dd>
<dt>quant</dt>
<dd><p>Numeric. Single number between 0 and 1 indicating the quantile 
used to calculate the cutoff. Default 0.99.</p></dd>
<dt>python</dt>
<dd><p>A logical value indicating whether to use singular-value 
decomposition from the python package torch.</p></dd>
<dt>store_perm</dt>
<dd><p>Logical. Whether permuted data should be stored in the CA 
object.
This implementation dramatically speeds up computation compared to `svd()` 
in R.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Returns the input "cacomp" object with "APL_score" component added.
APL_score contains a data frame with ranked rows, their score and their 
original row number.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The score is calculated by permuting the values of each row to determine the 
cutoff angle of the 99<!-- % quantile. -->
$$S_{alpha}(x,y)=x-\frac{y}{\tan\alpha}$$
By default the permutation is repeated 10 times, but for very large matrices 
this can be reduced.
If store_perm is TRUE the permuted data is stored in the cacomp object and 
can be used for future scoring.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Association Plots: Visualizing associations in high-dimensional 
correspondence analysis biplots <br>
Elzbieta Gralinska, Martin Vingron <br>
bioRxiv 2020.10.23.352096; doi: https://doi.org/10.1101/2020.10.23.352096</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Simulate counts</span></span>
<span class="r-in"><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span>, lambda <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">}</span>,</span>
<span class="r-in">               x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Run correspondence analysis.</span></span>
<span class="r-in"><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="cacomp.html">cacomp</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cnts</span>, princ_coords <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Parameter top is &gt;nrow(obj) and therefore ignored.</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Calculate APL coordinates:</span></span>
<span class="r-in"><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="apl_coords.html">apl_coords</a></span><span class="op">(</span><span class="va">ca</span>, group <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Rank genes by S-alpha score</span></span>
<span class="r-in"><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">apl_score</span><span class="op">(</span><span class="va">ca</span>, mat <span class="op">=</span> <span class="va">cnts</span><span class="op">)</span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   |                                                                              |                                                                      |   0%  |                                                                              |=======                                                               |  10%  |                                                                              |==============                                                        |  20%  |                                                                              |=====================                                                 |  30%  |                                                                              |============================                                          |  40%  |                                                                              |===================================                                   |  50%  |                                                                              |==========================================                            |  60%  |                                                                              |=================================================                     |  70%  |                                                                              |========================================================              |  80%  |                                                                              |===============================================================       |  90%  |                                                                              |======================================================================| 100%</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Elzbieta Gralinska, Clemens Kohl, Martin Vingron.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

