<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Correspondance Analysis — cacomp • APL</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Correspondance Analysis — cacomp"><meta property="og:description" content="`cacomp` performs correspondence analysis on a matrix or
Seurat/SingleCellExperiment object and returns the transformed data.
`cacomp.seurat` performs correspondence analysis on an assay from a Seurat 
container and stores the standardized coordinates of the columns (= cells) 
and the principal coordinates of the rows (= genes) as a DimReduc Object in 
the Seurat container.
`cacomp.SingleCellExperiment` performs correspondence analysis on an assay 
from a SingleCellExperiment and stores the standardized coordinates
 of the columns (= cells) and the principal coordinates of the rows 
 (= genes) as a matrix in the SingleCellExperiment container."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">APL</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.99.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/APL.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Correspondance Analysis</h1>
    
    <div class="hidden name"><code>cacomp.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>`cacomp` performs correspondence analysis on a matrix or
Seurat/SingleCellExperiment object and returns the transformed data.</p>
<p>`cacomp.seurat` performs correspondence analysis on an assay from a Seurat 
container and stores the standardized coordinates of the columns (= cells) 
and the principal coordinates of the rows (= genes) as a DimReduc Object in 
the Seurat container.</p>
<p>`cacomp.SingleCellExperiment` performs correspondence analysis on an assay 
from a SingleCellExperiment and stores the standardized coordinates
 of the columns (= cells) and the principal coordinates of the rows 
 (= genes) as a matrix in the SingleCellExperiment container.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="fu">cacomp</span><span class="op">(</span>
  <span class="va">obj</span>,
  coords <span class="op">=</span> <span class="cn">TRUE</span>,
  princ_coords <span class="op">=</span> <span class="fl">3</span>,
  python <span class="op">=</span> <span class="cn">FALSE</span>,
  dims <span class="op">=</span> <span class="cn">NULL</span>,
  top <span class="op">=</span> <span class="fl">5000</span>,
  inertia <span class="op">=</span> <span class="cn">TRUE</span>,
  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="co"># S4 method for matrix</span>
<span class="fu">cacomp</span><span class="op">(</span>
  <span class="va">obj</span>,
  coords <span class="op">=</span> <span class="cn">TRUE</span>,
  princ_coords <span class="op">=</span> <span class="fl">3</span>,
  python <span class="op">=</span> <span class="cn">FALSE</span>,
  dims <span class="op">=</span> <span class="cn">NULL</span>,
  top <span class="op">=</span> <span class="fl">5000</span>,
  inertia <span class="op">=</span> <span class="cn">TRUE</span>,
  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>
<span class="op">)</span>

<span class="co"># S4 method for Seurat</span>
<span class="fu">cacomp</span><span class="op">(</span>
  <span class="va">obj</span>,
  coords <span class="op">=</span> <span class="cn">TRUE</span>,
  princ_coords <span class="op">=</span> <span class="fl">3</span>,
  python <span class="op">=</span> <span class="cn">FALSE</span>,
  dims <span class="op">=</span> <span class="cn">NULL</span>,
  top <span class="op">=</span> <span class="fl">5000</span>,
  inertia <span class="op">=</span> <span class="cn">TRUE</span>,
  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>,
  assay <span class="op">=</span> <span class="fu">Seurat</span><span class="fu">::</span><span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/DefaultAssay.html" class="external-link">DefaultAssay</a></span><span class="op">(</span><span class="va">obj</span><span class="op">)</span>,
  slot <span class="op">=</span> <span class="st">"counts"</span>,
  return_input <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span>

<span class="co"># S4 method for SingleCellExperiment</span>
<span class="fu">cacomp</span><span class="op">(</span>
  <span class="va">obj</span>,
  coords <span class="op">=</span> <span class="cn">TRUE</span>,
  princ_coords <span class="op">=</span> <span class="fl">3</span>,
  python <span class="op">=</span> <span class="cn">FALSE</span>,
  dims <span class="op">=</span> <span class="cn">NULL</span>,
  top <span class="op">=</span> <span class="fl">5000</span>,
  inertia <span class="op">=</span> <span class="cn">TRUE</span>,
  rm_zeros <span class="op">=</span> <span class="cn">TRUE</span>,
  <span class="va">...</span>,
  assay <span class="op">=</span> <span class="st">"counts"</span>,
  return_input <span class="op">=</span> <span class="cn">FALSE</span>
<span class="op">)</span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>obj</dt>
<dd><p>A numeric matrix or Seurat/SingleCellExperiment object.
For sequencing a count matrix, gene expression values with genes in rows 
and samples/cells in columns.
Should contain row and column names.</p></dd>
<dt>coords</dt>
<dd><p>Logical. Indicates whether CA standard coordinates should be 
calculated. Default TRUE</p></dd>
<dt>princ_coords</dt>
<dd><p>Integer. Number indicating whether principal 
coordinates should be calculated for the rows (=1), columns (=2), 
both (=3) or none (=0).
Default 1.</p></dd>
<dt>python</dt>
<dd><p>A logical value indicating whether to use singular-value 
decomposition from the python package torch.
This implementation dramatically speeds up computation compared to `svd()` 
in R.</p></dd>
<dt>dims</dt>
<dd><p>Integer. Number of CA dimensions to retain. Default NULL 
(keeps all dimensions).</p></dd>
<dt>top</dt>
<dd><p>Integer. Number of most variable rows to retain. 
Default 5000. (set NULL to keep all).</p></dd>
<dt>inertia</dt>
<dd><p>Logical.. Whether total, row and column inertias should be 
calculated and returned. Default TRUE.</p></dd>
<dt>rm_zeros</dt>
<dd><p>Logical. Whether rows &amp; cols containing only 0s should be 
removed.
Keeping zero only rows/cols might lead to unexpected results. Default TRUE.</p></dd>
<dt>...</dt>
<dd><p>Other parameters</p></dd>
<dt>assay</dt>
<dd><p>Character. The assay from which extract the count matrix for 
SVD, e.g. "RNA" for Seurat objects or "counts"/"logcounts" for 
SingleCellExperiments.</p></dd>
<dt>slot</dt>
<dd><p>character. The slot of the Seurat assay. Default "counts".</p></dd>
<dt>return_input</dt>
<dd><p>Logical. If TRUE returns the input 
(SingleCellExperiment/Seurat object) with the CA results saved in the 
reducedDim/DimReduc slot "CA".
 Otherwise returns a "cacomp". Default FALSE.</p></dd>
</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>Returns a named list of class "cacomp" with components
U, V and D: The results from the SVD.
row_masses and col_masses: Row and columns masses.
top_rows: How many of the most variable rows were retained for the analysis.
tot_inertia, row_inertia and col_inertia: Only if inertia = TRUE.
Total, row and column inertia respectively.
If return_imput = TRUE with Seurat container: Returns input obj of class 
"Seurat" with a new Dimensional Reduction Object named "CA".
Standard coordinates of the cells are saved as embeddings,
the principal coordinates of the genes as loadings and
the singular values (= square root of principal intertias/eigenvalues)
are stored as stdev.
To recompute a regular "cacomp" object without rerunning cacomp use 
`as.cacomp()`.
If return_input =TRUE for SingleCellExperiment input returns a 
SingleCellExperiment object with a matrix of standardized coordinates of 
the columns in
reducedDim(obj, "CA"). Additionally, the matrix contains the following 
attributes:
"prin_coords_rows": Principal coordinates of the rows.
"singval": Singular values. For the explained inertia of each principal 
axis calculate singval^2.
"percInertia": Percent explained inertia of each principal axis.
To recompute a regular "cacomp" object from a SingleCellExperiment without 
rerunning cacomp use `as.cacomp()`.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The calculation is performed according to the work of Michael Greenacre. 
Singular value decomposition can be performed either with the base R 
function `svd` or preferably by the faster pytorch implementation 
(python = TRUE). When working with large matrices, CA coordinates and
principal coordinates should only be computed when needed to save 
computational time.</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Greenacre, M. Correspondence Analysis in Practice, Third Edition, 2017.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span class="co"># Simulate scRNAseq data.</span></span>
<span class="r-in"><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>cell_1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">5</span><span class="op">)</span>,</span>
<span class="r-in">                   cell_2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span>,</span>
<span class="r-in">                   cell_3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span class="r-in"><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Run correspondence analysis.</span></span>
<span class="r-in"><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span>obj <span class="op">=</span> <span class="va">cnts</span>, princ_coords <span class="op">=</span> <span class="fl">3</span>, top <span class="op">=</span> <span class="fl">5</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co">###########</span></span>
<span class="r-in"><span class="co"># Seurat  #</span></span>
<span class="r-in"><span class="co">###########</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Simulate counts</span></span>
<span class="r-in"><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span>, lambda <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">}</span>,</span>
<span class="r-in">                     x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Create Seurat object</span></span>
<span class="r-in"><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://mojaveazure.github.io/seurat-object/reference/CreateSeuratObject.html" class="external-link">CreateSeuratObject</a></span><span class="op">(</span>counts <span class="op">=</span> <span class="va">cnts</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span>Feature names cannot have underscores ('_'), replacing with dashes ('-')</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Run CA and save in dim. reduction slot</span></span>
<span class="r-in"><span class="va">seu</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">seu</span>, return_input <span class="op">=</span> <span class="cn">TRUE</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Parameter top is &gt;nrow(obj) and therefore ignored.</span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Run CA and return cacomp object</span></span>
<span class="r-in"><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">seu</span>, return_input <span class="op">=</span> <span class="cn">FALSE</span>, assay <span class="op">=</span> <span class="st">"RNA"</span>, slot <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> <span class="warning">Warning: </span></span>
<span class="r-wrn co"><span class="r-pr">#&gt;</span> Parameter top is &gt;nrow(obj) and therefore ignored.</span>
<span class="r-in"></span>
<span class="r-in"><span class="co">########################</span></span>
<span class="r-in"><span class="co"># SingleCellExperiment #</span></span>
<span class="r-in"><span class="co">########################</span></span>
<span class="r-in"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Simulate counts</span></span>
<span class="r-in"><span class="va">cnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/mapply.html" class="external-link">mapply</a></span><span class="op">(</span><span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html" class="external-link">rpois</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fl">500</span>, lambda <span class="op">=</span> <span class="va">x</span><span class="op">)</span><span class="op">}</span>,</span>
<span class="r-in">               x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, <span class="fl">50</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"cell_"</span>, <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">cnts</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"><span class="va">logcnts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html" class="external-link">log2</a></span><span class="op">(</span><span class="va">cnts</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># Create SingleCellExperiment object</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/SingleCellExperiment.html" class="external-link">SingleCellExperiment</a></span><span class="op">(</span>assays<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>counts<span class="op">=</span><span class="va">cnts</span>, logcounts<span class="op">=</span><span class="va">logcnts</span><span class="op">)</span><span class="op">)</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># run CA and save in dim. reduction slot.</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">sce</span>, return_input <span class="op">=</span> <span class="cn">TRUE</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span> <span class="co"># on counts</span></span>
<span class="r-in"><span class="va">sce</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">sce</span>, return_input <span class="op">=</span> <span class="cn">TRUE</span>, assay <span class="op">=</span> <span class="st">"logcounts"</span><span class="op">)</span> <span class="co"># on logcounts</span></span>
<span class="r-in"></span>
<span class="r-in"><span class="co"># run CA and return cacomp object.</span></span>
<span class="r-in"><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu">cacomp</span><span class="op">(</span><span class="va">sce</span>, return_input <span class="op">=</span> <span class="cn">FALSE</span>, assay <span class="op">=</span> <span class="st">"counts"</span><span class="op">)</span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Elzbieta Gralinska, Clemens Kohl, Martin Vingron.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.1.</p>
</div>

      </footer></div>

  


  

  </body></html>

