<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="APL">
<title>Analyzing data with APL • APL</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analyzing data with APL">
<meta property="og:description" content="APL">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">APL</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.7.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item">
  <a class="nav-link" href="../articles/APL.html">Get started</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav"></ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Analyzing data with APL</h1>
                        <h4 data-toc-skip class="author">Elzbieta
Gralinska</h4>
            <address class="author_afil">
      Max Planck Institute for Molecular Genetics, Berlin,
Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:gralinska@molgen.mpg.de" class="email">gralinska@molgen.mpg.de</a>
      </address>
                              <h4 data-toc-skip class="author">Clemens
Kohl</h4>
            <address class="author_afil">
      Max Planck Institute for Molecular Genetics, Berlin,
Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:kohl@molgen.mpg.de" class="email">kohl@molgen.mpg.de</a>
      </address>
                              <h4 data-toc-skip class="author">Martin
Vingron</h4>
            <address class="author_afil">
      Max Planck Institute for Molecular Genetics, Berlin,
Germany<br><a class="author_email" href="mailto:#"></a><a href="mailto:vingron@molgen.mpg.de" class="email">vingron@molgen.mpg.de</a>
      </address>
                  
      
      
      <div class="d-none name"><code>APL.Rmd</code></div>
    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>This package performs correspondence analysis (CA) and allows
      to identify cluster-specific genes using Association Plots (AP).
      Additionally, APL computes the cluster-specificity scores for all
      genes which allows to rank the genes by their specificity for a
      selected cell cluster of interest.</p>
    </div>
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>“APL” is a package developed for computation of Association Plots, a
method for visualization and analysis of single cell transcriptomics
data. The main focus of “APL” is the identification of genes
characteristic for individual clusters of cells from input data.</p>
<p>When working with <em>APL</em> package please cite:</p>
<pre><code>Gralinska, E., Kohl, C., Fadakar, B. S., &amp; Vingron, M. (2022). 
Visualizing Cluster-specific Genes from Single-cell Transcriptomics Data Using Association Plots. 
Journal of Molecular Biology, 434(11), 167525.</code></pre>
<p>A citation can also be obtained in R by running
<code>citation("APL")</code>. For a mathematical description of the
method, please refer to the manuscript.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the <em>APL</em> from Bioconductor, run:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://bioconductor.github.io/BiocManager/reference/install.html" class="external-link">install</a></span><span class="op">(</span><span class="st">"APL"</span><span class="op">)</span></span></code></pre></div>
<p>Alternatively the package can also be installed from GitHub:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://devtools.r-lib.org/" class="external-link">devtools</a></span><span class="op">)</span></span>
<span><span class="fu">install_github</span><span class="op">(</span><span class="st">"VingronLab/APL"</span><span class="op">)</span></span></code></pre></div>
<p>To additionally build the package vignette, run instead</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">install_github</span><span class="op">(</span><span class="st">"VingronLab/APL"</span>, build_vignettes <span class="op">=</span> <span class="cn">TRUE</span>, dependencies <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Building the vignette will however take considerable time.</p>
<div class="section level3">
<h3 id="pytorch-installation">pytorch installation<a class="anchor" aria-label="anchor" href="#pytorch-installation"></a>
</h3>
<p>In order to decrease the computation time of the singular value
decomposition (SVD), we highly recommend the installation of
<code>pytorch</code>. More information on the <code>pytorch</code>
installation is given below. Instead of installing <code>pytorch</code>,
users can also opt to use the R native SVD. For this, please use the
argument <code>python = FALSE</code> wherever applicable in this
vignette.</p>
<div class="section level4">
<h4 id="install-pytorch-with-reticulate">Install pytorch with reticulate<a class="anchor" aria-label="anchor" href="#install-pytorch-with-reticulate"></a>
</h4>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://rstudio.github.io/reticulate/" class="external-link">reticulate</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/install_miniconda.html" class="external-link">install_miniconda</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/use_python.html" class="external-link">use_condaenv</a></span><span class="op">(</span>condaenv <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/miniconda_path.html" class="external-link">miniconda_path</a></span><span class="op">(</span><span class="op">)</span>,<span class="st">"envs/r-reticulate"</span><span class="op">)</span>,</span>
<span>             required<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span>      </span>
<span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/conda-tools.html" class="external-link">conda_install</a></span><span class="op">(</span>envname <span class="op">=</span> <span class="st">"r-reticulate"</span>, packages <span class="op">=</span> <span class="st">"numpy"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rstudio.github.io/reticulate/reference/conda-tools.html" class="external-link">conda_install</a></span><span class="op">(</span>envname <span class="op">=</span> <span class="st">"r-reticulate"</span>, packages <span class="op">=</span> <span class="st">"pytorch"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="manually-install-pytorch-with-conda">Manually install pytorch with conda<a class="anchor" aria-label="anchor" href="#manually-install-pytorch-with-conda"></a>
</h4>
<p>To install <code>pytorch</code> please download the appropriate
Miniconda installer for your system from <a href="https://docs.conda.io/en/latest/miniconda.html" class="external-link">the conda
website</a>. Follow the installation instructions on their website and
make sure the R package <code>reticulate</code> is also installed before
proceeding. Once installed, list all available conda environments via
<br><code>conda info --envs</code> <br> One of the environments should
have <code>r-reticulate</code> in its name. Depending on where you
installed it and your system, the exact path might be different.
Activate the environment and install <code>pytorch</code> into it.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> activate ~/.local/share/r-miniconda/envs/r-reticulate <span class="co"># change path accordingly.</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install numpy</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="ex">conda</span> install pytorch</span></code></pre></div>
</div>
</div>
</div>
<div class="section level2">
<h2 id="preprocessing">Preprocessing<a class="anchor" aria-label="anchor" href="#preprocessing"></a>
</h2>
<div class="section level3">
<h3 id="setup">Setup<a class="anchor" aria-label="anchor" href="#setup"></a>
</h3>
<p>In this vignette we will use a small data set published by <a href="https://doi.org/10.1073/pnas.1507125112" class="external-link">Darmanis et
al. (2015)</a> consisting of 466 human adult cortical single cells
sequenced on the Fluidigm platform as an example. To obtain the data
necessary to follow the vignette we use the Bioconductor package <em><a href="https://bioconductor.org/packages/3.18/scRNAseq" class="external-link">scRNAseq</a></em>.</p>
<p>Besides the package <em>APL</em> we will use Bioconductor packages to
preprocess the data. Namely we will use <em><a href="https://bioconductor.org/packages/3.18/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>,
<em><a href="https://bioconductor.org/packages/3.18/scater" class="external-link">scater</a></em> and
<em><a href="https://bioconductor.org/packages/3.18/scran" class="external-link">scran</a></em>.
However, the preprocessing could equally be performed with the
single-cell RNA-seq analysis suite <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em>.</p>
<p>The preprocessing steps are performed according to the
recommendations published in <a href="https://bioconductor.org/books/release/OSCA/" class="external-link">Orchestrating
Single-Cell Analysis with Bioconductor</a> by Amezquita <em>et al.</em>
(2022). For more information about the rational behind them please refer
to the book.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://vingronlab.github.io/APL/">APL</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scRNAseq</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">SingleCellExperiment</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">scran</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="http://bioconductor.org/packages/scater/" class="external-link">scater</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1234</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="loading-the-data">Loading the data<a class="anchor" aria-label="anchor" href="#loading-the-data"></a>
</h3>
<p>We start with the loading and preprocessing of the Darmanis data.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">darmanis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scRNAseq/man/DarmanisBrainData.html" class="external-link">DarmanisBrainData</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; see ?scRNAseq and browseVignettes('scRNAseq') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="co">#&gt; see ?scRNAseq and browseVignettes('scRNAseq') for documentation</span></span>
<span><span class="co">#&gt; downloading 1 resources</span></span>
<span><span class="co">#&gt; retrieving 1 resource</span></span>
<span><span class="co">#&gt; loading from cache</span></span>
<span><span class="va">darmanis</span></span>
<span><span class="co">#&gt; class: SingleCellExperiment </span></span>
<span><span class="co">#&gt; dim: 22085 466 </span></span>
<span><span class="co">#&gt; metadata(0):</span></span>
<span><span class="co">#&gt; assays(1): counts</span></span>
<span><span class="co">#&gt; rownames(22085): 1/2-SBSRNA4 A1BG ... ZZZ3 tAKR</span></span>
<span><span class="co">#&gt; rowData names(0):</span></span>
<span><span class="co">#&gt; colnames(466): GSM1657871 GSM1657872 ... GSM1658365 GSM1658366</span></span>
<span><span class="co">#&gt; colData names(5): age c1.chip.id cell.type experiment_sample_name</span></span>
<span><span class="co">#&gt;   tissue</span></span>
<span><span class="co">#&gt; reducedDimNames(0):</span></span>
<span><span class="co">#&gt; mainExpName: NULL</span></span>
<span><span class="co">#&gt; altExpNames(0):</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="normalization-pca-clustering">Normalization, PCA &amp; Clustering<a class="anchor" aria-label="anchor" href="#normalization-pca-clustering"></a>
</h3>
<p>Association Plots from <em>APL</em> should be computed based on the
normalized expression data. Therefore, we first normalize the counts
from the Darmanis data and calculate both PCA and UMAP for
visualizations later.</p>
<p>For now, <em>APL</em> requires the data to be clustered beforehand.
The darmanis data comes already annotated, so we will use the cell types
stored in the <code>cell.type</code> metadata column instead of
performing a clustering.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span></span>
<span><span class="va">clust</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/quickCluster.html" class="external-link">quickCluster</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span> </span>
<span><span class="va">darmanis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/computeSumFactors.html" class="external-link">computeSumFactors</a></span><span class="op">(</span><span class="va">darmanis</span>, cluster<span class="op">=</span><span class="va">clust</span>, min.mean<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span></span>
<span><span class="va">darmanis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scuttle/man/logNormCounts.html" class="external-link">logNormCounts</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span></span>
<span></span>
<span><span class="va">dec</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/modelGeneVar.html" class="external-link">modelGeneVar</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span></span>
<span><span class="va">top_darmanis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/getTopHVGs.html" class="external-link">getTopHVGs</a></span><span class="op">(</span><span class="va">dec</span>, n<span class="op">=</span><span class="fl">5000</span><span class="op">)</span></span>
<span><span class="va">darmanis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scran/man/fixedPCA.html" class="external-link">fixedPCA</a></span><span class="op">(</span><span class="va">darmanis</span>, subset.row<span class="op">=</span><span class="va">top_darmanis</span><span class="op">)</span> </span>
<span><span class="va">darmanis</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/scater/man/runUMAP.html" class="external-link">runUMAP</a></span><span class="op">(</span><span class="va">darmanis</span>, dimred<span class="op">=</span><span class="st">"PCA"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span></span>
<span><span class="co">#&gt; Also defined by 'spam'</span></span>
<span><span class="co">#&gt; Found more than one class "dist" in cache; using the first, from namespace 'BiocGenerics'</span></span>
<span><span class="co">#&gt; Also defined by 'spam'</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">darmanis</span>, dimred<span class="op">=</span><span class="st">"UMAP"</span>, colour_by<span class="op">=</span><span class="st">"cell.type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="APL_files/figure-html/preprocess-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick start<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The fastest way to compute the Association Plot for a selected
cluster of cells from the input data is by using a wrapper function
<code><a href="../reference/runAPL.html">runAPL()</a></code>. <code><a href="../reference/runAPL.html">runAPL()</a></code> automates most of the
analysis steps for ease of use.</p>
<p>For example, to generate an Association Plot for the oligodendrocytes
we can use the following command:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/runAPL.html">runAPL</a></span><span class="op">(</span><span class="va">darmanis</span>,</span>
<span>       assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>       top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>       group <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">$</span><span class="va">cell.type</span> <span class="op">==</span> <span class="st">"oligodendrocytes"</span><span class="op">)</span>,</span>
<span>       type <span class="op">=</span> <span class="st">"ggplot"</span>,</span>
<span>       python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in rm_zeros(obj): Matrix contains rows with only 0s. These rows were</span></span>
<span><span class="co">#&gt; removed. If undesired set rm_zeros = FALSE.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  Using 74 dimensions. Subsetting.</span></span></code></pre></div>
<p><img src="APL_files/figure-html/runAPL-1.png" width="700"></p>
<pre><code>#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |======================================================================| 100%
#&gt; 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |======================================================================| 100%</code></pre>
<p>The generated Association Plot is computed based on the
log-normalized count matrix. By default <code>runAPL</code> uses the top
5,000 most variable genes in the data, but the data can be subset to any
number of genes by changing the value for the argument <code>top</code>.
The dimensionality of the CA is determined automatically by the elbow
rule described below (see <a href="#dim_reduc">here</a>). This default
behavior can be overriden by setting the dimensions manually (parameter
<code>dims</code>). The cluster-specificity score (<span class="math inline">\(S_\alpha\)</span>) for each gene is also
calculated (<code>score = TRUE</code>). In order to better explore the
data, <code>type</code> can be set to <code>"plotly"</code> to obtain an
interactive plot. <code>runAPL</code> has many arguments to further
customize the output and fine tune the calculations. Please refer to the
documentation (<code><a href="../reference/runAPL.html">?runAPL</a></code>) for more information. The following
sections in this vignette will discuss the choice of dimensionality and
the <span class="math inline">\(S_\alpha\)</span>-score.</p>
</div>
<div class="section level2">
<h2 id="step-by-step-way-of-computing-association-plots">Step-by-step way of computing Association Plots<a class="anchor" aria-label="anchor" href="#step-by-step-way-of-computing-association-plots"></a>
</h2>
<p>Alternatively, Association Plots can be computed step-by-step. This
allows to adjust the Association Plots to user’s needs. Below we explain
each step of the process of generating Association Plots.</p>
<div class="section level3">
<h3 id="correspondence-analysis">Correspondence Analysis<a class="anchor" aria-label="anchor" href="#correspondence-analysis"></a>
</h3>
<p>The first step of Association Plot computations is correspondence
analysis (CA). CA is a data dimensionality reduction method similar to
PCA, however it allows for a simultaneous embedding of both cells and
genes from the input data in the same space. In this example we perform
CA on the log-normalized count matrix of the darmanis brain data.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Computing CA on logcounts</span></span>
<span><span class="va">logcounts</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span></span>
<span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp.html">cacomp</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">logcounts</span>,</span>
<span>             top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>             python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in rm_zeros(obj): Matrix contains rows with only 0s. These rows were</span></span>
<span><span class="co">#&gt; removed. If undesired set rm_zeros = FALSE.</span></span>
<span></span>
<span><span class="co"># The above is equivalent to:</span></span>
<span><span class="co"># ca &lt;- cacomp(obj = darmanis,</span></span>
<span><span class="co">#              assay = "logcounts",</span></span>
<span><span class="co">#              top = 5000,</span></span>
<span><span class="co">#              python = TRUE)</span></span></code></pre></div>
<p>The function <code>cacomp</code> accepts as an input any matrix with
non-negative entries, be it a single-cell RNA-seq, bulk RNA-seq or other
data. For ease of use, <code>cacomp</code> accepts also <em><a href="https://bioconductor.org/packages/3.18/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
and <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em>
objects, however for these we additionally have to specify via the
<code>assay</code> and/or <code>slot</code> (for Seurat) parameter from
where to extract the data. Importantly, in order to ensure the
interpretability of the results <code>cacomp</code> (and related
functions such as <code>runAPL</code>) requires that the input matrix
contains both row and column names.</p>
<p>When performing a feature selection before CA, we can set the
argument <code>top</code> to the desired number of genes with the
highest variance across cells from the input data to retain for further
analysis. By default, only the top 5,000 most variable genes are kept as
a good compromise between computational time and keeping the most
relevant genes. If we want to ensure however that even marker genes of
smaller clusters are kept, we can increase the number of genes.
<!-- For this vignette we keep all genes in the data to ensure that also genes from smaller clusters such as the platelets are kept. -->
<!-- In many cases however 5,000 genes are a good compromise between computational time and keeping most relevant genes. --></p>
<p>The output of <code>cacomp</code> is an object of class
<code>cacomp</code>:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ca</span></span>
<span><span class="co">#&gt; cacomp object with 466 columns, 5000 rows and 465 dimensions.</span></span>
<span><span class="co">#&gt; Calc. standard coord.:  std_coords_rows, std_coords_cols</span></span>
<span><span class="co">#&gt; Calc. principal coord.: prin_coords_rows, prin_coords_cols</span></span>
<span><span class="co">#&gt; Calc. APL coord.:       </span></span>
<span><span class="co">#&gt; Explained inertia:      1.1% Dim1, 0.7% Dim2</span></span></code></pre></div>
<p>As can be seen in the summarized output, by default both types of
coordinates in the CA space (principal and standardized) are calculated.
Once the coordinates for the Association Plot are calculated, they will
also be shown in the output of <code>cacomp</code>. Slots are accessed
through an accessor function:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"std_coords_cols"</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">5</span><span class="op">]</span></span>
<span><span class="co">#&gt;                  Dim1       Dim2       Dim3        Dim4       Dim5</span></span>
<span><span class="co">#&gt; GSM1657871 -0.1067086 -0.1253297  0.5046485 -0.32461780  2.4262520</span></span>
<span><span class="co">#&gt; GSM1657872  1.5101357 -0.4299418 -0.1219273  0.07942823  0.2362747</span></span>
<span><span class="co">#&gt; GSM1657873 -0.2237619 -0.2610148  0.1955599  0.16963578  2.1477040</span></span>
<span><span class="co">#&gt; GSM1657874 -0.5680872 -0.1138251  0.4725071  0.04409559 -0.6708562</span></span>
<span><span class="co">#&gt; GSM1657875 -0.4739344 -0.2505648  0.4384626  0.10316896  2.8910899</span></span></code></pre></div>
<p>In the case of <em><a href="https://bioconductor.org/packages/3.18/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
and <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em>
objects, we can alternatively set <code>return_input = TRUE</code> to
get the input object back, with the CA results computed by “APL” and
stored in the appropriate slot for dimension reduction. This also allows
for using the plotting functions that come with these packages:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">darmanis</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp.html">cacomp</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">darmanis</span>,</span>
<span>               assay <span class="op">=</span> <span class="st">"logcounts"</span>,</span>
<span>               top <span class="op">=</span> <span class="fl">5000</span>,</span>
<span>               return_input <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>               python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in rm_zeros(obj): Matrix contains rows with only 0s. These rows were</span></span>
<span><span class="co">#&gt; removed. If undesired set rm_zeros = FALSE.</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">darmanis</span>, dimred<span class="op">=</span><span class="st">"CA"</span>,</span>
<span>               ncomponents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>,<span class="fl">2</span><span class="op">)</span>,</span>
<span>               colour_by<span class="op">=</span><span class="st">"cell.type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="APL_files/figure-html/ca_pbmc-1.png" width="700"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">darmanis</span>, dimred<span class="op">=</span><span class="st">"CA"</span>,</span>
<span>               ncomponents <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span>,<span class="fl">4</span><span class="op">)</span>,</span>
<span>               colour_by<span class="op">=</span><span class="st">"cell.type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="APL_files/figure-html/ca_pbmc-2.png" width="700"></p>
<p>However, some functions such as apl_coords() require information that
cannot be stored in the single-cell container objects. It is therefore
often easier to work with a <code>cacomp</code> object instead. We can
convert <em><a href="https://CRAN.R-project.org/package=Seurat" class="external-link">Seurat</a></em> or
<em><a href="https://bioconductor.org/packages/3.18/SingleCellExperiment" class="external-link">SingleCellExperiment</a></em>
objects which have CA results stored to a <code>cacomp</code> object
using the function <code><a href="../reference/as.cacomp.html">as.cacomp()</a></code>:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Converting the object pbmc to cacomp</span></span>
<span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as.cacomp.html">as.cacomp</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="dim_reduc">Reducing the number of CA dimensions<a class="anchor" aria-label="anchor" href="#dim_reduc"></a>
</h3>
<p>When working with high-dimensional data, after singular value
decomposition there will often be many dimensions which are representing
the noise in the data. In order to minimize the noise, it is generally
recommended to reduce the dimensionsionality of the data before
generating Association Plots.</p>
<p>The number of dimensions to retain can be computed using the function
<code>pick_dims</code>. This function offers three standard methods
which we implemented:</p>
<ul>
<li><p>elbow rule (<code>method = "elbow_rule"</code>) - the number of
dimensions to retain is calculated based on scree plots generated for
randomized data, and corresponds to a point in the plot where the band
of randomized singular values enters the band of the original singular
values,</p></li>
<li><p>80% rule (<code>method = "maj_inertia"</code>) - only those first
dimensions are retained which in total account for &gt;= 80% of total
inertia,</p></li>
<li><p>average rule (<code>method = "avg_inertia"</code>) - only those
dimensions are retained which account for more inertia than a single
dimension on average.</p></li>
</ul>
<p>Additionally, the user can compute a scree plot to choose the number
of dimensions by themselves:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pick_dims.html">pick_dims</a></span><span class="op">(</span><span class="va">ca</span>, method <span class="op">=</span> <span class="st">"scree_plot"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/lims.html" class="external-link">xlim</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">20</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning: Removed 445 rows containing missing values</span></span>
<span><span class="co">#&gt; (`position_stack()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 1 rows containing missing values (`geom_col()`).</span></span>
<span><span class="co">#&gt; Warning: Removed 445 rows containing missing values (`geom_line()`).</span></span></code></pre></div>
<p><img src="APL_files/figure-html/scree_plot-1.png" width="700"></p>
<p>In the scree plot above we can see that the first dimension explains
only ~1% of the total inertia and we observe the “jump” in the scree
plot at roughly 5 dimensions. The first dimensions however explain only
a small amount of the total inertia.</p>
<p>Here we compute the number of dimensions using the elbow rule. For
demonstration, only three data permutations are computed:</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pd</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/pick_dims.html">pick_dims</a></span><span class="op">(</span><span class="va">ca</span>,</span>
<span>                mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span>,</span>
<span>                method <span class="op">=</span> <span class="st">"elbow_rule"</span>,</span>
<span>                reps <span class="op">=</span> <span class="fl">3</span>, </span>
<span>                python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pd</span></span>
<span><span class="co">#&gt; [1] 74</span></span></code></pre></div>
<p>In this case the elbow rule leads to a much higher number of
dimensions.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute the amount of inertia explained by each of the dimensions</span></span>
<span><span class="va">D</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"D"</span><span class="op">)</span></span>
<span><span class="va">expl_inertia</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">D</span><span class="op">^</span><span class="fl">2</span><span class="op">/</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">D</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span><span class="op">)</span><span class="op">*</span><span class="fl">100</span></span>
<span></span>
<span><span class="co"># Compute the amount of intertia explained </span></span>
<span><span class="co"># by the number of dimensions defined by elbow rule</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">expl_inertia</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_len</a></span><span class="op">(</span><span class="va">pd</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 28.00469</span></span></code></pre></div>
<p>In this example the elbow rule suggests to keep 74 dimensions that
explain 28% of the total inertia from the data.</p>
<p>Finally, we can reduce the dimensionality of the data to the desired
number of dimensions:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/subset_dims.html">subset_dims</a></span><span class="op">(</span><span class="va">ca</span>, dims <span class="op">=</span> <span class="va">pd</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="association-plots">Association Plots<a class="anchor" aria-label="anchor" href="#association-plots"></a>
</h3>
<p>When working with single-cell transcriptomics data we are often
interested in which genes are associated to a cluster of cells. To
reveal such genes we can compute an Association Plot for a selected
cluster of cells. In the following example we want to generate an
Association Plot for the cluster of endothelial cells:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Specifying a cell cluster of interest</span></span>
<span><span class="va">endo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">$</span><span class="va">cell.type</span> <span class="op">==</span> <span class="st">"endothelial"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate Association Plot coordinates for endothelial cells</span></span>
<span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_coords.html">apl_coords</a></span><span class="op">(</span><span class="va">ca</span>, group <span class="op">=</span> <span class="va">endo</span><span class="op">)</span></span></code></pre></div>
<p>After computing the coordinates of genes and cells in the Association
Plot we are able to plot the results using the <code>apl</code>
function.</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># endothelial marker genes</span></span>
<span><span class="va">marker_genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"APOLD1"</span>, <span class="st">"TM4SF1"</span>, <span class="st">"SULT1B1"</span>, <span class="st">"ESM1"</span>, <span class="st">"SELE"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Plot APL</span></span>
<span><span class="fu"><a href="../reference/apl.html">apl</a></span><span class="op">(</span><span class="va">ca</span>,</span>
<span>    row_labs <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    rows_idx <span class="op">=</span> <span class="va">marker_genes</span>, </span>
<span>    type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span> <span class="co"># type = "plotly" for an interactive plot</span></span></code></pre></div>
<p><img src="APL_files/figure-html/apl_platelets_plot-1.png" width="700"></p>
<p>In the Association Plot all genes are represented by blue circles.
The further to the right a gene is located the more associated it is
with the chosen cluster of cells and the lower the y-axis value, the
more specific it is for the selected cluster. Additionally, it is
possible to highlight in the Association Plot any set of genes. In the
example above we highlighted five genes (APOLD1, TM4SF1, SULT1B1, ESM1,
SELE) which are known to be marker genes for endothelial cells. As we
can see in the plot, they are located in the right part of the plot,
which confirms their specificity for endothelial cells.</p>
<p>By default we plot only the genes in the Association Plot. To also
display the cells in the Association Plot, use the argument
<code>show_cols = TRUE</code>. This way we can identify other cells
which show similar expression profiles to the cells of interest. Cells
that belong to the cluster of interest will be colored in red, and all
remaining cells will be colored in violet. Furthermore, an interactive
plot in which you can hover over genes to see their name can be created
by setting <code>type = "plotly"</code>.</p>
</div>
<div class="section level3">
<h3 id="association-plots-with-the-s_alpha-scores">Association Plots with the <span class="math inline">\(S_\alpha\)</span>-scores<a class="anchor" aria-label="anchor" href="#association-plots-with-the-s_alpha-scores"></a>
</h3>
<p>The <span class="math inline">\(S_\alpha\)</span>-score allows us to
rank genes by their specificity for a selected cell cluster, and is
computed for each gene from the Association Plot separately. The higher
the <span class="math inline">\(S_\alpha\)</span>-score of a gene, the
more characteristic its expression for the investigated cell cluster.
The <span class="math inline">\(S_\alpha\)</span>-scores can be computed
using the <code>apl_score</code> function. To display the <span class="math inline">\(S_\alpha\)</span>-scores in the Association Plot,
we can use the argument <code>show_score = TRUE</code> in the
<code>apl</code> function:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Compute S-alpha score</span></span>
<span><span class="co"># For the calculation the input matrix is also required.</span></span>
<span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_score.html">apl_score</a></span><span class="op">(</span><span class="va">ca</span>,</span>
<span>                mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span>,</span>
<span>                reps <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/apl.html">apl</a></span><span class="op">(</span><span class="va">ca</span>,</span>
<span>    show_score <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>    type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span> </span></code></pre></div>
<p><img src="APL_files/figure-html/apl_plot_platelets-1.png" width="700"></p>
<p>By default, only genes that have a <span class="math inline">\(S_\alpha\)</span>-score larger than 0 are colored
as these tend to be genes of interest and we consider them as
cluster-specific genes. This cutoff can be easily changed through the
<code>score_cutoff</code> argument to <code><a href="../reference/apl.html">apl()</a></code>.</p>
<p>The <span class="math inline">\(S_\alpha\)</span>-scores are stored
in the <code>"APL_score"</code> slot and can be accessed as follows:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"APL_score"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt;             Rowname    Score Row_num Rank</span></span>
<span><span class="co">#&gt; PRKCDBP     PRKCDBP 2.816535    2390    1</span></span>
<span><span class="co">#&gt; IFITM1       IFITM1 2.725801    2081    2</span></span>
<span><span class="co">#&gt; OR2A20P     OR2A20P 2.633763    4888    3</span></span>
<span><span class="co">#&gt; OR4C46       OR4C46 2.633763    4889    4</span></span>
<span><span class="co">#&gt; LOC646268 LOC646268 2.633763      81    5</span></span>
<span><span class="co">#&gt; FOXC1         FOXC1 2.493272    2937    6</span></span></code></pre></div>
<p>To see the expression of genes with the highest <span class="math inline">\(S_\alpha\)</span>-scores (or any selected genes)
across all cell types from the data we can use plotting functions
provided by <em><a href="https://CRAN.R-project.org/package=scater" class="external-link">scater</a></em>:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">scores</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/cacomp_slot.html">cacomp_slot</a></span><span class="op">(</span><span class="va">ca</span>, <span class="st">"APL_score"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotExpression.html" class="external-link">plotExpression</a></span><span class="op">(</span><span class="va">darmanis</span>, </span>
<span>               features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">scores</span><span class="op">$</span><span class="va">Rowname</span>,<span class="fl">3</span><span class="op">)</span>,</span>
<span>               x <span class="op">=</span> <span class="st">"cell.type"</span>, </span>
<span>               colour_by <span class="op">=</span> <span class="st">"cell.type"</span><span class="op">)</span></span></code></pre></div>
<p><img src="APL_files/figure-html/seurat_apl-1.png" width="700"></p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/scater/man/plotReducedDim.html" class="external-link">plotReducedDim</a></span><span class="op">(</span><span class="va">darmanis</span>, </span>
<span>               dimred<span class="op">=</span><span class="st">"UMAP"</span>,</span>
<span>               colour_by<span class="op">=</span> <span class="va">scores</span><span class="op">$</span><span class="va">Rowname</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<p><img src="APL_files/figure-html/seurat_apl-2.png" width="700"></p>
<p>As expected, the 3 most highly scored genes are over-expressed in the
endothelial cells. Due to the small size of the data set and number of
cells in the cluster (only 20 out of 466 cells are endothelial cells)
some cluster specific genes are only expressed in a few cells. Most data
sets nowadays are significantly larger so this should not be a major
concern and it can further be mitigated by performing a more stringent
feature selection before CA.</p>
</div>
<div class="section level3">
<h3 id="visualization-of-ca">Visualization of CA<a class="anchor" aria-label="anchor" href="#visualization-of-ca"></a>
</h3>
<p>In addition to Association Plots “APL” produces also other forms of
the output. For instance, we can use “APL” to generate a two- and
three-dimensional correspondence analysis projection of the data. The
so-called biplot visualizes both cells and genes from the input data and
can be created using the function <code>ca_biplot</code>. Alternatively,
a three-dimensional data projection plot can be generated using the
function <code>ca_3Dplot</code>. To generate such biplots a
<code>cacomp</code> object is required.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Specifying a cell cluster of interest</span></span>
<span><span class="va">endo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">$</span><span class="va">cell.type</span> <span class="op">==</span> <span class="st">"endothelial"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Creating a static plot</span></span>
<span><span class="fu"><a href="../reference/ca_biplot.html">ca_biplot</a></span><span class="op">(</span><span class="va">ca</span>, col_labels <span class="op">=</span> <span class="va">endo</span>, type <span class="op">=</span> <span class="st">"ggplot"</span><span class="op">)</span></span></code></pre></div>
<p><img src="APL_files/figure-html/biplot-1.png" width="700"></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span></span>
<span><span class="co"># Creating an interactive plot</span></span>
<span><span class="co"># ca_biplot(ca, type = "plotly", col_labels = platelets)</span></span>
<span></span>
<span><span class="co"># 3D plot</span></span>
<span><span class="co"># ca_3Dplot(ca, col_labels = platelets)</span></span></code></pre></div>
<p>The above described plots give us a quick overview of the first 2
dimensions of the data (more dimensions can be plotted). As shown in the
commented-out code, to interactively explore the projection of the data
<code>type = "plotly"</code> can be set.</p>
</div>
</div>
<div class="section level2">
<h2 id="apl-and-go-enrichment-analysis">APL and GO enrichment analysis<a class="anchor" aria-label="anchor" href="#apl-and-go-enrichment-analysis"></a>
</h2>
<p>After computing an Association Plot and identifying a set of genes
specific for a selected cluster of cells we might be interested in
conducting a Gene Ontology (GO) enrichment analysis of the identified
gene set. To conduct a GO enrichment analysis of microglia specific
genes as idenitfied using an Association Plot, we first need to compute
the coordinates of the genes in the Association Plot for microglia
cells, as well as the <span class="math inline">\(S_\alpha\)</span>-score for each gene:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Get indices of microglia cells</span></span>
<span><span class="va">microglia</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">$</span><span class="va">cell.type</span> <span class="op">==</span> <span class="st">"microglia"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Calculate Association Plot coordinates of the genes and the $S_\alpha$-scores</span></span>
<span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_coords.html">apl_coords</a></span><span class="op">(</span><span class="va">ca</span>, group <span class="op">=</span> <span class="va">microglia</span><span class="op">)</span></span>
<span></span>
<span><span class="va">ca</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_score.html">apl_score</a></span><span class="op">(</span><span class="va">ca</span>,</span>
<span>                mat <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/pkg/SingleCellExperiment/man/assays.html" class="external-link">logcounts</a></span><span class="op">(</span><span class="va">darmanis</span><span class="op">)</span>,</span>
<span>                reps <span class="op">=</span> <span class="fl">5</span>,</span>
<span>                python <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Now we can conduct GO enrichment analysis as implemented in the
package <em><a href="https://bioconductor.org/packages/3.18/topGO" class="external-link">topGO</a></em> using
the most cluster-specific genes from the Association Plot. By default we
use all genes with an <span class="math inline">\(S_\alpha\)</span>-score higher than 0, but the
cutoff may have to be adjusted depending on the dataset. In the example
below we restrict it to genes with a <span class="math inline">\(S_\alpha\)</span>-score higher than 1 to restrict
it to truly significant genes. In case that no <span class="math inline">\(S_\alpha\)</span>-scores were calculated, one can
also choose to use the <code>ngenes</code> (by default 1000) genes with
the highest x-coordinates by setting <code>use_coords = TRUE</code>.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">enr</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/apl_topGO.html">apl_topGO</a></span><span class="op">(</span><span class="va">ca</span>,</span>
<span>                 ontology <span class="op">=</span> <span class="st">"BP"</span>,</span>
<span>                 organism <span class="op">=</span> <span class="st">"hs"</span>,</span>
<span>                 score_cutoff <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enr</span><span class="op">)</span></span>
<span><span class="co">#&gt;        GO.ID</span></span>
<span><span class="co">#&gt; 1 GO:0045087</span></span>
<span><span class="co">#&gt; 2 GO:0019886</span></span>
<span><span class="co">#&gt; 3 GO:0032755</span></span>
<span><span class="co">#&gt; 4 GO:0032760</span></span>
<span><span class="co">#&gt; 5 GO:0002250</span></span>
<span><span class="co">#&gt; 6 GO:0002503</span></span>
<span><span class="co">#&gt;                                                                                Term</span></span>
<span><span class="co">#&gt; 1                                                            innate immune response</span></span>
<span><span class="co">#&gt; 2 antigen processing and presentation of exogenous peptide antigen via MHC class II</span></span>
<span><span class="co">#&gt; 3                                   positive regulation of interleukin-6 production</span></span>
<span><span class="co">#&gt; 4                           positive regulation of tumor necrosis factor production</span></span>
<span><span class="co">#&gt; 5                                                          adaptive immune response</span></span>
<span><span class="co">#&gt; 6                        peptide antigen assembly with MHC class II protein complex</span></span>
<span><span class="co">#&gt;   Annotated Significant Expected raw.p.value</span></span>
<span><span class="co">#&gt; 1       278          57    17.67     6.5e-12</span></span>
<span><span class="co">#&gt; 2        15          11     0.95     5.8e-11</span></span>
<span><span class="co">#&gt; 3        39          15     2.48     4.4e-09</span></span>
<span><span class="co">#&gt; 4        45          16     2.86     5.0e-09</span></span>
<span><span class="co">#&gt; 5       166          41    10.55     2.5e-08</span></span>
<span><span class="co">#&gt; 6        11           8     0.70     3.3e-08</span></span></code></pre></div>
<p>The function <code><a href="../reference/plot_enrichment.html">plot_enrichment()</a></code> was implemented to
visualize the <code>topGO</code> results in form of a dotplot.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plot_enrichment.html">plot_enrichment</a></span><span class="op">(</span><span class="va">enr</span><span class="op">)</span></span></code></pre></div>
<p><img src="APL_files/figure-html/topGO_plot-1.png" width="700">
Microglia cells are innate immune cells of the brain and as such the
most highly scored genes are enriched in gene sets related to the immune
response and microglia specific gene sets as one would expect.</p>
</div>
<div class="section level2">
<h2 class="unnumbered" id="session-info">Session info<a class="anchor" aria-label="anchor" href="#session-info"></a>
</h2>
<pre><code><span><span class="co">#&gt; R version 4.3.2 (2023-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats4    stats     graphics  grDevices utils     datasets  methods  </span></span>
<span><span class="co">#&gt; [8] base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt;  [1] SparseM_1.81                org.Hs.eg.db_3.18.0        </span></span>
<span><span class="co">#&gt;  [3] AnnotationDbi_1.64.1        scater_1.30.1              </span></span>
<span><span class="co">#&gt;  [5] ggplot2_3.4.4               scran_1.30.2               </span></span>
<span><span class="co">#&gt;  [7] scuttle_1.12.0              scRNAseq_2.16.0            </span></span>
<span><span class="co">#&gt;  [9] SingleCellExperiment_1.24.0 SummarizedExperiment_1.32.0</span></span>
<span><span class="co">#&gt; [11] Biobase_2.62.0              GenomicRanges_1.54.1       </span></span>
<span><span class="co">#&gt; [13] GenomeInfoDb_1.38.5         IRanges_2.36.0             </span></span>
<span><span class="co">#&gt; [15] S4Vectors_0.40.2            BiocGenerics_0.48.1        </span></span>
<span><span class="co">#&gt; [17] MatrixGenerics_1.14.0       matrixStats_1.2.0          </span></span>
<span><span class="co">#&gt; [19] APL_1.7.1                   BiocStyle_2.30.0           </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] fs_1.6.3                      ProtGenerics_1.34.0          </span></span>
<span><span class="co">#&gt;   [3] spatstat.sparse_3.0-3         bitops_1.0-7                 </span></span>
<span><span class="co">#&gt;   [5] httr_1.4.7                    RColorBrewer_1.1-3           </span></span>
<span><span class="co">#&gt;   [7] tools_4.3.2                   sctransform_0.4.1            </span></span>
<span><span class="co">#&gt;   [9] utf8_1.2.4                    R6_2.5.1                     </span></span>
<span><span class="co">#&gt;  [11] lazyeval_0.2.2                uwot_0.1.16                  </span></span>
<span><span class="co">#&gt;  [13] withr_3.0.0                   sp_2.1-3                     </span></span>
<span><span class="co">#&gt;  [15] prettyunits_1.2.0             gridExtra_2.3                </span></span>
<span><span class="co">#&gt;  [17] progressr_0.14.0              cli_3.6.2                    </span></span>
<span><span class="co">#&gt;  [19] textshaping_0.3.7             spatstat.explore_3.2-6       </span></span>
<span><span class="co">#&gt;  [21] fastDummies_1.7.3             labeling_0.4.3               </span></span>
<span><span class="co">#&gt;  [23] sass_0.4.8                    topGO_2.54.0                 </span></span>
<span><span class="co">#&gt;  [25] Seurat_5.0.1                  spatstat.data_3.0-4          </span></span>
<span><span class="co">#&gt;  [27] ggridges_0.5.6                pbapply_1.7-2                </span></span>
<span><span class="co">#&gt;  [29] pkgdown_2.0.7                 Rsamtools_2.18.0             </span></span>
<span><span class="co">#&gt;  [31] systemfonts_1.0.5             parallelly_1.36.0            </span></span>
<span><span class="co">#&gt;  [33] limma_3.58.1                  RSQLite_2.3.5                </span></span>
<span><span class="co">#&gt;  [35] FNN_1.1.4                     generics_0.1.3               </span></span>
<span><span class="co">#&gt;  [37] BiocIO_1.12.0                 ica_1.0-3                    </span></span>
<span><span class="co">#&gt;  [39] spatstat.random_3.2-2         dplyr_1.1.4                  </span></span>
<span><span class="co">#&gt;  [41] GO.db_3.18.0                  Matrix_1.6-5                 </span></span>
<span><span class="co">#&gt;  [43] ggbeeswarm_0.7.2              fansi_1.0.6                  </span></span>
<span><span class="co">#&gt;  [45] abind_1.4-5                   lifecycle_1.0.4              </span></span>
<span><span class="co">#&gt;  [47] edgeR_4.0.14                  yaml_2.3.8                   </span></span>
<span><span class="co">#&gt;  [49] SparseArray_1.2.3             BiocFileCache_2.10.1         </span></span>
<span><span class="co">#&gt;  [51] Rtsne_0.17                    grid_4.3.2                   </span></span>
<span><span class="co">#&gt;  [53] blob_1.2.4                    dqrng_0.3.2                  </span></span>
<span><span class="co">#&gt;  [55] promises_1.2.1                ExperimentHub_2.10.0         </span></span>
<span><span class="co">#&gt;  [57] crayon_1.5.2                  miniUI_0.1.1.1               </span></span>
<span><span class="co">#&gt;  [59] lattice_0.21-9                beachmat_2.18.0              </span></span>
<span><span class="co">#&gt;  [61] cowplot_1.1.3                 GenomicFeatures_1.54.3       </span></span>
<span><span class="co">#&gt;  [63] KEGGREST_1.42.0               metapod_1.10.1               </span></span>
<span><span class="co">#&gt;  [65] pillar_1.9.0                  knitr_1.45                   </span></span>
<span><span class="co">#&gt;  [67] rjson_0.2.21                  future.apply_1.11.1          </span></span>
<span><span class="co">#&gt;  [69] codetools_0.2-19              leiden_0.4.3.1               </span></span>
<span><span class="co">#&gt;  [71] glue_1.7.0                    data.table_1.15.0            </span></span>
<span><span class="co">#&gt;  [73] vctrs_0.6.5                   png_0.1-8                    </span></span>
<span><span class="co">#&gt;  [75] spam_2.10-0                   org.Mm.eg.db_3.18.0          </span></span>
<span><span class="co">#&gt;  [77] gtable_0.3.4                  cachem_1.0.8                 </span></span>
<span><span class="co">#&gt;  [79] xfun_0.41                     S4Arrays_1.2.0               </span></span>
<span><span class="co">#&gt;  [81] mime_0.12                     survival_3.5-7               </span></span>
<span><span class="co">#&gt;  [83] statmod_1.5.0                 bluster_1.12.0               </span></span>
<span><span class="co">#&gt;  [85] interactiveDisplayBase_1.40.0 ellipsis_0.3.2               </span></span>
<span><span class="co">#&gt;  [87] fitdistrplus_1.1-11           ROCR_1.0-11                  </span></span>
<span><span class="co">#&gt;  [89] nlme_3.1-163                  bit64_4.0.5                  </span></span>
<span><span class="co">#&gt;  [91] progress_1.2.3                filelock_1.0.3               </span></span>
<span><span class="co">#&gt;  [93] RcppAnnoy_0.0.22              rprojroot_2.0.4              </span></span>
<span><span class="co">#&gt;  [95] bslib_0.6.1                   irlba_2.3.5.1                </span></span>
<span><span class="co">#&gt;  [97] vipor_0.4.7                   KernSmooth_2.23-22           </span></span>
<span><span class="co">#&gt;  [99] colorspace_2.1-0              DBI_1.2.1                    </span></span>
<span><span class="co">#&gt; [101] tidyselect_1.2.0              bit_4.0.5                    </span></span>
<span><span class="co">#&gt; [103] compiler_4.3.2                curl_5.2.0                   </span></span>
<span><span class="co">#&gt; [105] graph_1.80.0                  BiocNeighbors_1.20.2         </span></span>
<span><span class="co">#&gt; [107] xml2_1.3.6                    desc_1.4.3                   </span></span>
<span><span class="co">#&gt; [109] DelayedArray_0.28.0           plotly_4.10.4                </span></span>
<span><span class="co">#&gt; [111] bookdown_0.37                 rtracklayer_1.62.0           </span></span>
<span><span class="co">#&gt; [113] scales_1.3.0                  lmtest_0.9-40                </span></span>
<span><span class="co">#&gt; [115] rappdirs_0.3.3                stringr_1.5.1                </span></span>
<span><span class="co">#&gt; [117] digest_0.6.34                 goftest_1.2-3                </span></span>
<span><span class="co">#&gt; [119] spatstat.utils_3.0-4          rmarkdown_2.25               </span></span>
<span><span class="co">#&gt; [121] XVector_0.42.0                htmltools_0.5.7              </span></span>
<span><span class="co">#&gt; [123] pkgconfig_2.0.3               sparseMatrixStats_1.14.0     </span></span>
<span><span class="co">#&gt; [125] highr_0.10                    dbplyr_2.4.0                 </span></span>
<span><span class="co">#&gt; [127] fastmap_1.1.1                 ensembldb_2.26.0             </span></span>
<span><span class="co">#&gt; [129] rlang_1.1.3                   htmlwidgets_1.6.4            </span></span>
<span><span class="co">#&gt; [131] shiny_1.8.0                   DelayedMatrixStats_1.24.0    </span></span>
<span><span class="co">#&gt; [133] farver_2.1.1                  jquerylib_0.1.4              </span></span>
<span><span class="co">#&gt; [135] zoo_1.8-12                    jsonlite_1.8.8               </span></span>
<span><span class="co">#&gt; [137] BiocParallel_1.36.0           BiocSingular_1.18.0          </span></span>
<span><span class="co">#&gt; [139] RCurl_1.98-1.14               magrittr_2.0.3               </span></span>
<span><span class="co">#&gt; [141] GenomeInfoDbData_1.2.11       dotCall64_1.1-1              </span></span>
<span><span class="co">#&gt; [143] patchwork_1.2.0               munsell_0.5.0                </span></span>
<span><span class="co">#&gt; [145] Rcpp_1.0.12                   viridis_0.6.5                </span></span>
<span><span class="co">#&gt; [147] reticulate_1.35.0.9000        stringi_1.8.3                </span></span>
<span><span class="co">#&gt; [149] zlibbioc_1.48.0               MASS_7.3-60                  </span></span>
<span><span class="co">#&gt; [151] AnnotationHub_3.10.0          plyr_1.8.9                   </span></span>
<span><span class="co">#&gt; [153] parallel_4.3.2                listenv_0.9.1                </span></span>
<span><span class="co">#&gt; [155] ggrepel_0.9.5                 deldir_2.0-2                 </span></span>
<span><span class="co">#&gt; [157] Biostrings_2.70.2             splines_4.3.2                </span></span>
<span><span class="co">#&gt; [159] tensor_1.5                    hms_1.1.3                    </span></span>
<span><span class="co">#&gt; [161] locfit_1.5-9.8                igraph_2.0.1.1               </span></span>
<span><span class="co">#&gt; [163] spatstat.geom_3.2-8           RcppHNSW_0.6.0               </span></span>
<span><span class="co">#&gt; [165] ScaledMatrix_1.10.0           reshape2_1.4.4               </span></span>
<span><span class="co">#&gt; [167] biomaRt_2.58.2                BiocVersion_3.18.1           </span></span>
<span><span class="co">#&gt; [169] XML_3.99-0.16.1               evaluate_0.23                </span></span>
<span><span class="co">#&gt; [171] SeuratObject_5.0.1            BiocManager_1.30.22          </span></span>
<span><span class="co">#&gt; [173] httpuv_1.6.14                 RANN_2.6.1                   </span></span>
<span><span class="co">#&gt; [175] tidyr_1.3.1                   purrr_1.0.2                  </span></span>
<span><span class="co">#&gt; [177] polyclip_1.10-6               future_1.33.1                </span></span>
<span><span class="co">#&gt; [179] scattermore_1.2               rsvd_1.0.5                   </span></span>
<span><span class="co">#&gt; [181] xtable_1.8-4                  restfulr_0.0.15              </span></span>
<span><span class="co">#&gt; [183] AnnotationFilter_1.26.0       RSpectra_0.16-1              </span></span>
<span><span class="co">#&gt; [185] later_1.3.2                   viridisLite_0.4.2            </span></span>
<span><span class="co">#&gt; [187] ragg_1.2.7                    tibble_3.2.1                 </span></span>
<span><span class="co">#&gt; [189] beeswarm_0.4.0                memoise_2.0.1                </span></span>
<span><span class="co">#&gt; [191] GenomicAlignments_1.38.2      cluster_2.1.4                </span></span>
<span><span class="co">#&gt; [193] globals_0.16.2                here_1.0.1</span></span></code></pre>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Clemens Kohl, Elzbieta Gralinska, Martin Vingron.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
